// @alfresco/adf-core/esm5/card-view/components/card-view-item-dispatcher/card-view-item-dispatcher.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as tslib_1 from "tslib";
import { Component, ComponentFactoryResolver, Input, ViewChild } from '@angular/core';
import { CardItemTypeService } from '../../services/card-item-types.service';
import { CardViewContentProxyDirective } from '../../directives/card-view-content-proxy.directive';
var CardViewItemDispatcherComponent = /** @class */ (function () {
    function CardViewItemDispatcherComponent(cardItemTypeService, resolver) {
        var _this = this;
        this.cardItemTypeService = cardItemTypeService;
        this.resolver = resolver;
        this.displayEmpty = true;
        this.loaded = false;
        this.componentReference = null;
        /** @type {?} */
        var dynamicLifecycleMethods = [
            'ngOnInit',
            'ngDoCheck',
            'ngAfterContentInit',
            'ngAfterContentChecked',
            'ngAfterViewInit',
            'ngAfterViewChecked',
            'ngOnDestroy'
        ];
        dynamicLifecycleMethods.forEach(function (dynamicLifecycleMethod) {
            _this[dynamicLifecycleMethod] = _this.proxy.bind(_this, dynamicLifecycleMethod);
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    CardViewItemDispatcherComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        if (!this.loaded) {
            this.loadComponent();
            this.loaded = true;
        }
        Object.keys(changes)
            .map(function (changeName) { return [changeName, changes[changeName]]; })
            .forEach(function (_a) {
            var _b = tslib_1.__read(_a, 2), inputParamName = _b[0], simpleChange = _b[1];
            _this.componentReference.instance[inputParamName] = simpleChange.currentValue;
        });
        this.proxy('ngOnChanges', changes);
    };
    /**
     * @return {?}
     */
    CardViewItemDispatcherComponent.prototype.loadComponent = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var factoryClass = this.cardItemTypeService.resolveComponentType(this.property);
        /** @type {?} */
        var factory = this.resolver.resolveComponentFactory(factoryClass);
        this.componentReference = this.content.viewContainerRef.createComponent(factory);
        this.componentReference.instance.editable = this.editable;
        this.componentReference.instance.property = this.property;
        this.componentReference.instance.displayEmpty = this.displayEmpty;
    };
    /**
     * @param {?} methodName
     * @param {...?} args
     * @return {?}
     */
    CardViewItemDispatcherComponent.prototype.proxy = /**
     * @param {?} methodName
     * @param {...?} args
     * @return {?}
     */
    function (methodName) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (this.componentReference.instance[methodName]) {
            this.componentReference.instance[methodName].apply(this.componentReference.instance, args);
        }
    };
    CardViewItemDispatcherComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-card-view-item-dispatcher',
                    template: '<ng-template adf-card-view-content-proxy></ng-template>'
                }] }
    ];
    /** @nocollapse */
    CardViewItemDispatcherComponent.ctorParameters = function () { return [
        { type: CardItemTypeService },
        { type: ComponentFactoryResolver }
    ]; };
    CardViewItemDispatcherComponent.propDecorators = {
        property: [{ type: Input }],
        editable: [{ type: Input }],
        displayEmpty: [{ type: Input }],
        content: [{ type: ViewChild, args: [CardViewContentProxyDirective,] }]
    };
    return CardViewItemDispatcherComponent;
}());
export { CardViewItemDispatcherComponent };
if (false) {
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.property;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.editable;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.displayEmpty;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.content;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.loaded;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.componentReference;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.ngOnInit;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.ngDoCheck;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.cardItemTypeService;
    /** @type {?} */
    CardViewItemDispatcherComponent.prototype.resolver;
}
