// @alfresco/adf-core/esm5/viewer/directives/viewer-extension.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ContentChild, Directive, Input, TemplateRef } from '@angular/core';
import { ViewerComponent } from '../components/viewer.component';
var ViewerExtensionDirective = /** @class */ (function () {
    function ViewerExtensionDirective(viewerComponent) {
        this.viewerComponent = viewerComponent;
    }
    /**
     * @return {?}
     */
    ViewerExtensionDirective.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.templateModel = { template: this.template, isVisible: false };
        this.viewerComponent.extensionTemplates.push(this.templateModel);
        this.viewerComponent.extensionChange.subscribe(function (fileExtension) {
            _this.templateModel.isVisible = _this.isVisible(fileExtension);
        });
        if (this.supportedExtensions instanceof Array) {
            this.supportedExtensions.forEach(function (extension) {
                _this.viewerComponent.externalExtensions.push(extension);
            });
        }
    };
    /**
     * check if the current extension in the viewer is compatible with this extension checking against supportedExtensions
     */
    /**
     * check if the current extension in the viewer is compatible with this extension checking against supportedExtensions
     * @param {?} fileExtension
     * @return {?}
     */
    ViewerExtensionDirective.prototype.isVisible = /**
     * check if the current extension in the viewer is compatible with this extension checking against supportedExtensions
     * @param {?} fileExtension
     * @return {?}
     */
    function (fileExtension) {
        /** @type {?} */
        var supportedExtension;
        if (this.supportedExtensions && (this.supportedExtensions instanceof Array)) {
            supportedExtension = this.supportedExtensions.find(function (extension) {
                return extension.toLowerCase() === fileExtension;
            });
        }
        return !!supportedExtension;
    };
    ViewerExtensionDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'adf-viewer-extension'
                },] }
    ];
    /** @nocollapse */
    ViewerExtensionDirective.ctorParameters = function () { return [
        { type: ViewerComponent }
    ]; };
    ViewerExtensionDirective.propDecorators = {
        template: [{ type: ContentChild, args: [TemplateRef,] }],
        urlFileContent: [{ type: Input }],
        extension: [{ type: Input }],
        supportedExtensions: [{ type: Input }]
    };
    return ViewerExtensionDirective;
}());
export { ViewerExtensionDirective };
if (false) {
    /** @type {?} */
    ViewerExtensionDirective.prototype.template;
    /** @type {?} */
    ViewerExtensionDirective.prototype.urlFileContent;
    /** @type {?} */
    ViewerExtensionDirective.prototype.extension;
    /** @type {?} */
    ViewerExtensionDirective.prototype.supportedExtensions;
    /** @type {?} */
    ViewerExtensionDirective.prototype.templateModel;
    /** @type {?} */
    ViewerExtensionDirective.prototype.viewerComponent;
}
