// @alfresco/adf-content-services/esm2015/search/components/search-check-list/search-check-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation } from '@angular/core';
import { SearchFilterList } from '../search-filter/models/search-filter-list.model';
/**
 * @record
 */
export function SearchListOption() { }
/** @type {?} */
SearchListOption.prototype.name;
/** @type {?} */
SearchListOption.prototype.value;
/** @type {?} */
SearchListOption.prototype.checked;
export class SearchCheckListComponent {
    constructor() {
        this.operator = 'OR';
        this.pageSize = 5;
        this.options = new SearchFilterList();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.settings) {
            this.operator = this.settings["operator"] || 'OR';
            this.pageSize = this.settings["pageSize"] || 5;
            if (this.settings["options"] && this.settings["options"].length > 0) {
                this.options = new SearchFilterList(this.settings["options"], this.pageSize);
            }
        }
    }
    /**
     * @return {?}
     */
    reset() {
        this.options.items.forEach(opt => {
            opt.checked = false;
        });
        if (this.id && this.context) {
            this.context.queryFragments[this.id] = '';
            this.context.update();
        }
    }
    /**
     * @param {?} event
     * @param {?} option
     * @return {?}
     */
    changeHandler(event, option) {
        option.checked = event.checked;
        this.flush();
    }
    /**
     * @return {?}
     */
    flush() {
        /** @type {?} */
        const checkedValues = this.options.items
            .filter(option => option.checked)
            .map(option => option.value);
        /** @type {?} */
        const query = checkedValues.join(` ${this.operator} `);
        if (this.id && this.context) {
            this.context.queryFragments[this.id] = query;
            this.context.update();
        }
    }
}
SearchCheckListComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-search-check-list',
                template: "<div class=\"checklist\">\n    <mat-checkbox\n        *ngFor=\"let option of options\"\n        [checked]=\"option.checked\"\n        (change)=\"changeHandler($event, option)\">\n        {{ option.name | translate }}\n    </mat-checkbox>\n</div>\n\n<div class=\"facet-buttons\" *ngIf=\"options.fitsPage\">\n    <button mat-button color=\"primary\" (click)=\"reset()\">\n        {{ 'SEARCH.FILTER.ACTIONS.CLEAR-ALL' | translate }}\n    </button>\n</div>\n\n<div class=\"facet-buttons\" *ngIf=\"!options.fitsPage\">\n    <button mat-icon-button\n        title=\"{{ 'SEARCH.FILTER.ACTIONS.CLEAR-ALL' | translate }}\"\n        (click)=\"reset()\">\n        <mat-icon>clear</mat-icon>\n    </button>\n    <button mat-icon-button\n        *ngIf=\"options.canShowLessItems\"\n        title=\"{{ 'SEARCH.FILTER.ACTIONS.SHOW-LESS' | translate }}\"\n        (click)=\"options.showLessItems()\">\n        <mat-icon>keyboard_arrow_up</mat-icon>\n    </button>\n    <button mat-icon-button\n        *ngIf=\"options.canShowMoreItems\"\n        title=\"{{ 'SEARCH.FILTER.ACTIONS.SHOW-MORE' | translate }}\"\n        (click)=\"options.showMoreItems()\">\n        <mat-icon>keyboard_arrow_down</mat-icon>\n    </button>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                host: { class: 'adf-search-check-list' },
                styles: [".adf-search-check-list{display:flex;flex-direction:column}"]
            }] }
];
/** @nocollapse */
SearchCheckListComponent.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    SearchCheckListComponent.prototype.id;
    /** @type {?} */
    SearchCheckListComponent.prototype.settings;
    /** @type {?} */
    SearchCheckListComponent.prototype.context;
    /** @type {?} */
    SearchCheckListComponent.prototype.options;
    /** @type {?} */
    SearchCheckListComponent.prototype.operator;
    /** @type {?} */
    SearchCheckListComponent.prototype.pageSize;
}
