// @alfresco/adf-core/esm2015/templates/error-content/error-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ChangeDetectionStrategy, Input, ViewEncapsulation } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { TranslationService } from '../../services/translation.service';
export class ErrorContentComponent {
    /**
     * @param {?} route
     * @param {?} router
     * @param {?} translateService
     */
    constructor(route, router, translateService) {
        this.route = route;
        this.router = router;
        this.translateService = translateService;
        /**
         * Target URL for the secondary button.
         */
        this.secondaryButtonUrl = 'report-issue';
        /**
         * Target URL for the return button.
         */
        this.returnButtonUrl = '/';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.route) {
            this.route.params.forEach((params) => {
                if (params['id'] && !this.errorCode) {
                    this.errorCode = params['id'];
                    /** @type {?} */
                    let unknown = '';
                    this.translateService.get('ERROR_CONTENT.' + this.errorCode + '.TITLE').subscribe((errorTranslation) => {
                        unknown = errorTranslation;
                    });
                    if (unknown === 'ERROR_CONTENT.' + this.errorCode + '.TITLE') {
                        this.errorCode = 'UNKNOWN';
                    }
                }
            });
        }
    }
    /**
     * @return {?}
     */
    getTranslations() {
        this.hasSecondButton = this.translateService.instant('ERROR_CONTENT.' + this.errorCode + '.SECONDARY_BUTTON.TEXT') ? true : false;
    }
    /**
     * @return {?}
     */
    ngAfterContentChecked() {
        this.getTranslations();
    }
    /**
     * @return {?}
     */
    onSecondButton() {
        this.router.navigate(['/' + this.secondaryButtonUrl]);
    }
    /**
     * @return {?}
     */
    onReturnButton() {
        this.router.navigate(['/' + this.returnButtonUrl]);
    }
}
ErrorContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-error-content',
                template: "<div class=\"adf-error-content\">\n    <p class=\"adf-error-content-code\">\n        {{ errorCode }}\n    </p>\n    <div class=\"adf-error-content-shadow\"></div>\n    <p class=\"adf-error-content-title\">\n        {{ 'ERROR_CONTENT.' + errorCode + '.TITLE' | translate }}\n    </p>\n    <p class=\"adf-error-content-description\">\n        {{ 'ERROR_CONTENT.' + errorCode + '.DESCRIPTION' | translate }}\n    </p>\n    <div class=\"adf-error-content-buttons\">\n        <a a id=\"adf-secondary-button\" mat-raised-button color=\"primary\"\n            *ngIf=\"hasSecondButton\" (click)=\"onSecondButton()\"\n            class=\"adf-error-content-description-link\">\n            {{ 'ERROR_CONTENT.' + errorCode + '.SECONDARY_BUTTON.TEXT' | translate | uppercase }}\n        </a>\n        <a id=\"adf-return-button\" mat-raised-button color=\"primary\" (click)=\"onReturnButton()\">\n            {{ 'ERROR_CONTENT.' + this.errorCode + '.RETURN_BUTTON.TEXT' | translate | uppercase }}\n        </a>\n    </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                host: { class: 'adf-error-content' },
                styles: [""]
            }] }
];
/** @nocollapse */
ErrorContentComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: Router },
    { type: TranslationService }
];
ErrorContentComponent.propDecorators = {
    secondaryButtonUrl: [{ type: Input }],
    returnButtonUrl: [{ type: Input }],
    errorCode: [{ type: Input }]
};
if (false) {
    /**
     * Target URL for the secondary button.
     * @type {?}
     */
    ErrorContentComponent.prototype.secondaryButtonUrl;
    /**
     * Target URL for the return button.
     * @type {?}
     */
    ErrorContentComponent.prototype.returnButtonUrl;
    /**
     * Error code associated with this error.
     * @type {?}
     */
    ErrorContentComponent.prototype.errorCode;
    /** @type {?} */
    ErrorContentComponent.prototype.hasSecondButton;
    /** @type {?} */
    ErrorContentComponent.prototype.route;
    /** @type {?} */
    ErrorContentComponent.prototype.router;
    /** @type {?} */
    ErrorContentComponent.prototype.translateService;
}
