// @alfresco/adf-core/esm2015/form/components/widgets/tabs/tabs.widget.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
export class TabsWidgetComponent {
    constructor() {
        this.tabs = [];
        this.formTabChanged = new EventEmitter();
        this.visibleTabs = [];
    }
    /**
     * @return {?}
     */
    hasTabs() {
        return this.tabs && this.tabs.length > 0;
    }
    /**
     * @return {?}
     */
    ngAfterContentChecked() {
        this.filterVisibleTabs();
    }
    /**
     * @return {?}
     */
    filterVisibleTabs() {
        this.visibleTabs = this.tabs.filter(tab => {
            return tab.isVisible;
        });
    }
    /**
     * @param {?} field
     * @return {?}
     */
    tabChanged(field) {
        this.formTabChanged.emit(field);
    }
}
TabsWidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'tabs-widget',
                template: "<div *ngIf=\"hasTabs()\" class=\"alfresco-tabs-widget\">\n    <mat-tab-group>\n        <mat-tab *ngFor=\"let tab of visibleTabs\" [label]=\"tab.title\">\n            <div *ngFor=\"let field of tab.fields\">\n                <adf-form-field [field]=\"field.field\"></adf-form-field>\n             </div>\n        </mat-tab>\n    </mat-tab-group>\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
TabsWidgetComponent.propDecorators = {
    tabs: [{ type: Input }],
    formTabChanged: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    TabsWidgetComponent.prototype.tabs;
    /** @type {?} */
    TabsWidgetComponent.prototype.formTabChanged;
    /** @type {?} */
    TabsWidgetComponent.prototype.visibleTabs;
}
