// @alfresco/adf-core/esm5/login/components/login-dialog-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation, ViewChild, Output, EventEmitter } from '@angular/core';
import { LoginComponent } from './login.component';
var LoginDialogPanelComponent = /** @class */ (function () {
    function LoginDialogPanelComponent() {
        this.success = new EventEmitter();
    }
    /**
     * @return {?}
     */
    LoginDialogPanelComponent.prototype.submitForm = /**
     * @return {?}
     */
    function () {
        this.login.submit();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    LoginDialogPanelComponent.prototype.onLoginSuccess = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.success.emit(event);
    };
    /**
     * @return {?}
     */
    LoginDialogPanelComponent.prototype.isValid = /**
     * @return {?}
     */
    function () {
        return this.login && this.login.form ? this.login.form.valid : false;
    };
    LoginDialogPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-login-dialog-panel',
                    template: "<div>\n    <adf-login #adfLogin\n               [showRememberMe]=\"false\"\n               [showLoginActions]=\"false\"\n               [backgroundImageUrl]=\"''\"\n               (success)=\"onLoginSuccess($event)\">\n        <login-header><ng-template></ng-template></login-header>\n        <login-footer><ng-template></ng-template></login-footer>\n    </adf-login>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [""]
                }] }
    ];
    LoginDialogPanelComponent.propDecorators = {
        success: [{ type: Output }],
        login: [{ type: ViewChild, args: ['adfLogin',] }]
    };
    return LoginDialogPanelComponent;
}());
export { LoginDialogPanelComponent };
if (false) {
    /** @type {?} */
    LoginDialogPanelComponent.prototype.success;
    /** @type {?} */
    LoginDialogPanelComponent.prototype.login;
}
