// @alfresco/adf-content-services/esm5/document-list/components/content-column/content-column-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { LogService } from '@alfresco/adf-core';
import { Component } from '@angular/core';
import { DocumentListComponent } from './../document-list.component';
var ContentColumnListComponent = /** @class */ (function () {
    function ContentColumnListComponent(documentList, logService) {
        this.documentList = documentList;
        this.logService = logService;
        this.logService.log('ContentColumnListComponent is deprecated starting with 1.7.0 and may be removed in future versions. Use DataColumnListComponent instead.');
    }
    /**
     * Registers column model within the parent document list component.
     * @param column Column definition model to register.
     */
    /**
     * Registers column model within the parent document list component.
     * @param {?} column Column definition model to register.
     * @return {?}
     */
    ContentColumnListComponent.prototype.registerColumn = /**
     * Registers column model within the parent document list component.
     * @param {?} column Column definition model to register.
     * @return {?}
     */
    function (column) {
        if (this.documentList && column) {
            /** @type {?} */
            var columns = this.documentList.data.getColumns();
            columns.push(column);
            return true;
        }
        return false;
    };
    ContentColumnListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'content-columns',
                    template: ''
                }] }
    ];
    /** @nocollapse */
    ContentColumnListComponent.ctorParameters = function () { return [
        { type: DocumentListComponent },
        { type: LogService }
    ]; };
    return ContentColumnListComponent;
}());
export { ContentColumnListComponent };
if (false) {
    /** @type {?} */
    ContentColumnListComponent.prototype.documentList;
    /** @type {?} */
    ContentColumnListComponent.prototype.logService;
}
