// @alfresco/adf-content-services/esm5/document-list/models/content-action.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as tslib_1 from "tslib";
var ContentActionModel = /** @class */ (function () {
    function ContentActionModel(obj) {
        this.disableWithNoPermission = false;
        this.disabled = false;
        this.visible = true;
        if (obj) {
            this.icon = obj.icon;
            this.title = obj.title;
            this.handler = obj.handler;
            this.execute = obj.execute;
            this.target = obj.target;
            this.permission = obj.permission;
            this.disableWithNoPermission = obj.disableWithNoPermission;
            if (obj.hasOwnProperty('disabled')) {
                this.disabled = obj.disabled;
            }
            if (obj.hasOwnProperty('visible')) {
                this.visible = obj.visible;
            }
        }
    }
    return ContentActionModel;
}());
export { ContentActionModel };
if (false) {
    /** @type {?} */
    ContentActionModel.prototype.icon;
    /** @type {?} */
    ContentActionModel.prototype.title;
    /** @type {?} */
    ContentActionModel.prototype.handler;
    /** @type {?} */
    ContentActionModel.prototype.execute;
    /** @type {?} */
    ContentActionModel.prototype.target;
    /** @type {?} */
    ContentActionModel.prototype.permission;
    /** @type {?} */
    ContentActionModel.prototype.disableWithNoPermission;
    /** @type {?} */
    ContentActionModel.prototype.disabled;
    /** @type {?} */
    ContentActionModel.prototype.visible;
}
/** @enum {string} */
var ContentActionTarget = {
    Document: 'document',
    Folder: 'folder',
    All: 'all',
};
export { ContentActionTarget };
/** @typedef {?} */
var ContentActionHandler;
export { ContentActionHandler };
var DocumentActionModel = /** @class */ (function (_super) {
    tslib_1.__extends(DocumentActionModel, _super);
    function DocumentActionModel(json) {
        var _this = _super.call(this, json) || this;
        _this.target = 'document';
        return _this;
    }
    return DocumentActionModel;
}(ContentActionModel));
export { DocumentActionModel };
var FolderActionModel = /** @class */ (function (_super) {
    tslib_1.__extends(FolderActionModel, _super);
    function FolderActionModel(json) {
        var _this = _super.call(this, json) || this;
        _this.target = 'folder';
        return _this;
    }
    return FolderActionModel;
}(ContentActionModel));
export { FolderActionModel };
