// @alfresco/adf-core/esm5/card-view/components/card-view/card-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, Input } from '@angular/core';
var CardViewComponent = /** @class */ (function () {
    function CardViewComponent() {
        /**
         * Toggles whether or not to show empty items in non-editable mode.
         */
        this.displayEmpty = true;
    }
    CardViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-card-view',
                    template: "<div class=\"adf-property-list\">\n    <div *ngFor=\"let property of properties\">\n        <div [attr.data-automation-id]=\"'header-'+property.key\" class=\"adf-property\">\n            <adf-card-view-item-dispatcher\n                [property]=\"property\"\n                [editable]=\"editable\"\n                [displayEmpty]=\"displayEmpty\">\n            </adf-card-view-item-dispatcher>\n        </div>\n    </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    CardViewComponent.propDecorators = {
        properties: [{ type: Input }],
        editable: [{ type: Input }],
        displayEmpty: [{ type: Input }]
    };
    return CardViewComponent;
}());
export { CardViewComponent };
if (false) {
    /**
     * (**required**) Items to show in the card view.
     * @type {?}
     */
    CardViewComponent.prototype.properties;
    /**
     * Toggles whether or not the items can be edited.
     * @type {?}
     */
    CardViewComponent.prototype.editable;
    /**
     * Toggles whether or not to show empty items in non-editable mode.
     * @type {?}
     */
    CardViewComponent.prototype.displayEmpty;
}
