// @alfresco/adf-core/esm5/form/components/widgets/container/container.widget.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation } from '@angular/core';
import { FormService } from './../../../services/form.service';
import { baseHost, WidgetComponent } from './../widget.component';
import { ContainerWidgetComponentModel } from './container.widget.model';
var ContainerWidgetComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ContainerWidgetComponent, _super);
    function ContainerWidgetComponent(formService) {
        var _this = _super.call(this, formService) || this;
        _this.formService = formService;
        return _this;
    }
    /**
     * @return {?}
     */
    ContainerWidgetComponent.prototype.onExpanderClicked = /**
     * @return {?}
     */
    function () {
        if (this.content && this.content.isCollapsible()) {
            this.content.isExpanded = !this.content.isExpanded;
        }
    };
    /**
     * @return {?}
     */
    ContainerWidgetComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.field) {
            this.content = new ContainerWidgetComponentModel(this.field);
        }
    };
    Object.defineProperty(ContainerWidgetComponent.prototype, "fields", {
        /**
         * Serializes column fields
         */
        get: /**
         * Serializes column fields
         * @return {?}
         */
        function () {
            /** @type {?} */
            var fields = [];
            /** @type {?} */
            var rowContainsElement = true;
            /** @type {?} */
            var rowIndex = 0;
            while (rowContainsElement) {
                rowContainsElement = false;
                for (var i = 0; i < this.content.columns.length; i++) {
                    /** @type {?} */
                    var field = this.content.columns[i].fields[rowIndex];
                    if (field) {
                        rowContainsElement = true;
                    }
                    fields.push(field);
                }
                rowIndex++;
            }
            return fields;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Calculate the column width based on the numberOfColumns and current field's colspan property
     *
     * @param field
     */
    /**
     * Calculate the column width based on the numberOfColumns and current field's colspan property
     *
     * @param {?} field
     * @return {?}
     */
    ContainerWidgetComponent.prototype.getColumnWith = /**
     * Calculate the column width based on the numberOfColumns and current field's colspan property
     *
     * @param {?} field
     * @return {?}
     */
    function (field) {
        /** @type {?} */
        var colspan = field ? field.colspan : 1;
        return (100 / this.content.json.numberOfColumns) * colspan + '%';
    };
    ContainerWidgetComponent.decorators = [
        { type: Component, args: [{
                    selector: 'container-widget',
                    template: "<div [hidden]=\"!content?.isGroup()\" class=\"container-widget__header\">\n    <h4 class=\"container-widget__header-text\" id=\"container-header\"\n        [class.collapsible]=\"content?.isCollapsible()\">\n        <button *ngIf=\"content?.isCollapsible()\"\n                mat-icon-button\n                class=\"mdl-button--icon\"\n                (click)=\"onExpanderClicked()\">\n            <mat-icon>{{ content?.isExpanded ? 'expand_more' : 'expand_less' }}</mat-icon>\n        </button>\n        <span (click)=\"onExpanderClicked()\" id=\"container-header-label\">{{content.name}}</span>\n    </h4>\n</div>\n\n<section class=\"grid-list\" *ngIf=\"content?.isExpanded\">\n    <div class=\"grid-list-item\" *ngFor=\"let field of fields\" [style.width]=\"getColumnWith(field)\">\n        <adf-form-field *ngIf=\"field\" [field]=\"field\"></adf-form-field>\n    </div>\n</section>\n\n\n",
                    host: baseHost,
                    encapsulation: ViewEncapsulation.None,
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ContainerWidgetComponent.ctorParameters = function () { return [
        { type: FormService }
    ]; };
    return ContainerWidgetComponent;
}(WidgetComponent));
export { ContainerWidgetComponent };
if (false) {
    /** @type {?} */
    ContainerWidgetComponent.prototype.content;
    /** @type {?} */
    ContainerWidgetComponent.prototype.formService;
}
