// @alfresco/adf-core/esm5/services/apps-process.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { from, throwError } from 'rxjs';
import { AlfrescoApiService } from './alfresco-api.service';
import { LogService } from './log.service';
import { map, catchError } from 'rxjs/operators';
var AppsProcessService = /** @class */ (function () {
    function AppsProcessService(apiService, logService) {
        this.apiService = apiService;
        this.logService = logService;
    }
    /**
     * Gets a list of deployed apps for this user.
     * @returns The list of deployed apps
     */
    /**
     * Gets a list of deployed apps for this user.
     * @return {?} The list of deployed apps
     */
    AppsProcessService.prototype.getDeployedApplications = /**
     * Gets a list of deployed apps for this user.
     * @return {?} The list of deployed apps
     */
    function () {
        var _this = this;
        return from(this.apiService.getInstance().activiti.appsApi.getAppDefinitions())
            .pipe(map(function (response) { return (response.data); }), catchError(function (err) { return _this.handleError(err); }));
    };
    /**
     * Gets a list of deployed apps for this user, where the app name is `name`.
     * @param name Name of the app
     * @returns The list of deployed apps
     */
    /**
     * Gets a list of deployed apps for this user, where the app name is `name`.
     * @param {?} name Name of the app
     * @return {?} The list of deployed apps
     */
    AppsProcessService.prototype.getDeployedApplicationsByName = /**
     * Gets a list of deployed apps for this user, where the app name is `name`.
     * @param {?} name Name of the app
     * @return {?} The list of deployed apps
     */
    function (name) {
        var _this = this;
        return from(this.apiService.getInstance().activiti.appsApi.getAppDefinitions())
            .pipe(map(function (response) { return (response.data.find(function (app) { return app.name === name; })); }), catchError(function (err) { return _this.handleError(err); }));
    };
    /**
     * Gets the details for a specific app ID number.
     * @param appId ID of the target app
     * @returns Details of the app
     */
    /**
     * Gets the details for a specific app ID number.
     * @param {?} appId ID of the target app
     * @return {?} Details of the app
     */
    AppsProcessService.prototype.getApplicationDetailsById = /**
     * Gets the details for a specific app ID number.
     * @param {?} appId ID of the target app
     * @return {?} Details of the app
     */
    function (appId) {
        var _this = this;
        return from(this.apiService.getInstance().activiti.appsApi.getAppDefinitions())
            .pipe(map(function (response) { return response.data.find(function (app) { return app.id === appId; }); }), catchError(function (err) { return _this.handleError(err); }));
    };
    /**
     * @param {?} error
     * @return {?}
     */
    AppsProcessService.prototype.handleError = /**
     * @param {?} error
     * @return {?}
     */
    function (error) {
        this.logService.error(error);
        return throwError(error || 'Server error');
    };
    AppsProcessService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AppsProcessService.ctorParameters = function () { return [
        { type: AlfrescoApiService },
        { type: LogService }
    ]; };
    return AppsProcessService;
}());
export { AppsProcessService };
if (false) {
    /** @type {?} */
    AppsProcessService.prototype.apiService;
    /** @type {?} */
    AppsProcessService.prototype.logService;
}
