// @alfresco/adf-core/esm5/services/auth-guard-ecm.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { AuthenticationService } from './authentication.service';
import { AppConfigService, AppConfigValues } from '../app-config/app-config.service';
var AuthGuardEcm = /** @class */ (function () {
    function AuthGuardEcm(authService, router, appConfig) {
        this.authService = authService;
        this.router = router;
        this.appConfig = appConfig;
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuardEcm.prototype.canActivate = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        return this.checkLogin(state.url);
    };
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuardEcm.prototype.canActivateChild = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        return this.canActivate(route, state);
    };
    /**
     * @param {?} redirectUrl
     * @return {?}
     */
    AuthGuardEcm.prototype.checkLogin = /**
     * @param {?} redirectUrl
     * @return {?}
     */
    function (redirectUrl) {
        if (this.authService.isEcmLoggedIn()) {
            return true;
        }
        if (!this.authService.isOauth() || this.isOAuthWithoutSilentLogin()) {
            this.authService.setRedirect({ provider: 'ECM', url: redirectUrl });
            /** @type {?} */
            var pathToLogin = this.getRouteDestinationForLogin();
            this.router.navigate(['/' + pathToLogin]);
        }
        return false;
    };
    /**
     * @return {?}
     */
    AuthGuardEcm.prototype.isOAuthWithoutSilentLogin = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var oauth = this.appConfig.get(AppConfigValues.OAUTHCONFIG, null);
        return this.authService.isOauth() && oauth.silentLogin === false;
    };
    /**
     * @return {?}
     */
    AuthGuardEcm.prototype.getRouteDestinationForLogin = /**
     * @return {?}
     */
    function () {
        return this.appConfig &&
            this.appConfig.get(AppConfigValues.LOGIN_ROUTE) ?
            this.appConfig.get(AppConfigValues.LOGIN_ROUTE) : 'login';
    };
    AuthGuardEcm.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AuthGuardEcm.ctorParameters = function () { return [
        { type: AuthenticationService },
        { type: Router },
        { type: AppConfigService }
    ]; };
    return AuthGuardEcm;
}());
export { AuthGuardEcm };
if (false) {
    /** @type {?} */
    AuthGuardEcm.prototype.authService;
    /** @type {?} */
    AuthGuardEcm.prototype.router;
    /** @type {?} */
    AuthGuardEcm.prototype.appConfig;
}
