// @alfresco/adf-core/esm5/services/deleted-nodes-api.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { from, of } from 'rxjs';
import { AlfrescoApiService } from './alfresco-api.service';
import { UserPreferencesService } from './user-preferences.service';
import { catchError } from 'rxjs/operators';
var DeletedNodesApiService = /** @class */ (function () {
    function DeletedNodesApiService(apiService, preferences) {
        this.apiService = apiService;
        this.preferences = preferences;
    }
    Object.defineProperty(DeletedNodesApiService.prototype, "nodesApi", {
        get: /**
         * @return {?}
         */
        function () {
            return this.apiService.getInstance().core.nodesApi;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Gets a list of nodes in the trash.
     * @param options Options for JSAPI call
     * @returns List of nodes in the trash
     */
    /**
     * Gets a list of nodes in the trash.
     * @param {?=} options Options for JSAPI call
     * @return {?} List of nodes in the trash
     */
    DeletedNodesApiService.prototype.getDeletedNodes = /**
     * Gets a list of nodes in the trash.
     * @param {?=} options Options for JSAPI call
     * @return {?} List of nodes in the trash
     */
    function (options) {
        /** @type {?} */
        var defaultOptions = {
            include: ['path', 'properties'],
            maxItems: this.preferences.paginationSize,
            skipCount: 0
        };
        /** @type {?} */
        var queryOptions = Object.assign(defaultOptions, options);
        /** @type {?} */
        var promise = this.nodesApi.getDeletedNodes(queryOptions);
        return from(promise).pipe(catchError(function (err) { return of(err); }));
    };
    DeletedNodesApiService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    DeletedNodesApiService.ctorParameters = function () { return [
        { type: AlfrescoApiService },
        { type: UserPreferencesService }
    ]; };
    return DeletedNodesApiService;
}());
export { DeletedNodesApiService };
if (false) {
    /** @type {?} */
    DeletedNodesApiService.prototype.apiService;
    /** @type {?} */
    DeletedNodesApiService.prototype.preferences;
}
