// @alfresco/adf-core/esm5/services/dynamic-component-mapper.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { getType } from './get-type';
/**
 * @record
 */
export function DynamicComponentModel() { }
/** @type {?} */
DynamicComponentModel.prototype.type;
/** @typedef {?} */
var DynamicComponentResolveFunction;
export { DynamicComponentResolveFunction };
var DynamicComponentResolver = /** @class */ (function () {
    function DynamicComponentResolver() {
    }
    /**
     * @param {?} type
     * @return {?}
     */
    DynamicComponentResolver.fromType = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return getType(type);
    };
    return DynamicComponentResolver;
}());
export { DynamicComponentResolver };
/**
 * @abstract
 */
var /**
 * @abstract
 */
DynamicComponentMapper = /** @class */ (function () {
    function DynamicComponentMapper() {
        this.defaultValue = undefined;
        this.types = {};
    }
    /**
     * Gets the currently active ComponentTypeResolver function for a field type.
     * @param type The type whose resolver you want
     * @param defaultValue Default type returned for types that are not yet mapped
     * @returns Resolver function
     */
    /**
     * Gets the currently active ComponentTypeResolver function for a field type.
     * @param {?} type The type whose resolver you want
     * @param {?=} defaultValue Default type returned for types that are not yet mapped
     * @return {?} Resolver function
     */
    DynamicComponentMapper.prototype.getComponentTypeResolver = /**
     * Gets the currently active ComponentTypeResolver function for a field type.
     * @param {?} type The type whose resolver you want
     * @param {?=} defaultValue Default type returned for types that are not yet mapped
     * @return {?} Resolver function
     */
    function (type, defaultValue) {
        if (defaultValue === void 0) { defaultValue = this.defaultValue; }
        if (type) {
            return this.types[type] || DynamicComponentResolver.fromType(defaultValue);
        }
        return DynamicComponentResolver.fromType(defaultValue);
    };
    /**
     * Sets or optionally replaces a ComponentTypeResolver function for a field type.
     * @param type The type whose resolver you want to set
     * @param resolver The new resolver function
     * @param override The new resolver will only replace an existing one if this parameter is true
     */
    /**
     * Sets or optionally replaces a ComponentTypeResolver function for a field type.
     * @param {?} type The type whose resolver you want to set
     * @param {?} resolver The new resolver function
     * @param {?=} override The new resolver will only replace an existing one if this parameter is true
     * @return {?}
     */
    DynamicComponentMapper.prototype.setComponentTypeResolver = /**
     * Sets or optionally replaces a ComponentTypeResolver function for a field type.
     * @param {?} type The type whose resolver you want to set
     * @param {?} resolver The new resolver function
     * @param {?=} override The new resolver will only replace an existing one if this parameter is true
     * @return {?}
     */
    function (type, resolver, override) {
        if (override === void 0) { override = true; }
        if (!type) {
            throw new Error("type is null or not defined");
        }
        if (!resolver) {
            throw new Error("resolver is null or not defined");
        }
        /** @type {?} */
        var existing = this.types[type];
        if (existing && !override) {
            throw new Error("already mapped, use override option if you intend replacing existing mapping.");
        }
        this.types[type] = resolver;
    };
    /**
     * Finds the component type that is needed to render a form field.
     * @param model Form field model for the field to render
     * @param defaultValue Default type returned for field types that are not yet mapped.
     * @returns Component type
     */
    /**
     * Finds the component type that is needed to render a form field.
     * @param {?} model Form field model for the field to render
     * @param {?=} defaultValue Default type returned for field types that are not yet mapped.
     * @return {?} Component type
     */
    DynamicComponentMapper.prototype.resolveComponentType = /**
     * Finds the component type that is needed to render a form field.
     * @param {?} model Form field model for the field to render
     * @param {?=} defaultValue Default type returned for field types that are not yet mapped.
     * @return {?} Component type
     */
    function (model, defaultValue) {
        if (defaultValue === void 0) { defaultValue = this.defaultValue; }
        if (model) {
            /** @type {?} */
            var resolver = this.getComponentTypeResolver(model.type, defaultValue);
            return resolver(model);
        }
        return defaultValue;
    };
    return DynamicComponentMapper;
}());
/**
 * @abstract
 */
export { DynamicComponentMapper };
if (false) {
    /** @type {?} */
    DynamicComponentMapper.prototype.defaultValue;
    /** @type {?} */
    DynamicComponentMapper.prototype.types;
}
