// @alfresco/adf-core/esm5/services/login-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { MatDialog } from '@angular/material';
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { LoginDialogComponent } from '../login/components/login-dialog.component';
var LoginDialogService = /** @class */ (function () {
    function LoginDialogService(dialog) {
        this.dialog = dialog;
    }
    /**
     * Opens a dialog to choose a file to upload.
     * @param action Name of the action to show in the title
     * @param contentEntry Item to upload
     * @returns Information about the chosen file(s)
     */
    /**
     * Opens a dialog to choose a file to upload.
     * @param {?} actionName
     * @param {?} title
     * @return {?} Information about the chosen file(s)
     */
    LoginDialogService.prototype.openLogin = /**
     * Opens a dialog to choose a file to upload.
     * @param {?} actionName
     * @param {?} title
     * @return {?} Information about the chosen file(s)
     */
    function (actionName, title) {
        /** @type {?} */
        var logged = new Subject();
        logged.subscribe({
            complete: this.close.bind(this)
        });
        /** @type {?} */
        var data = {
            title: title,
            actionName: actionName,
            logged: logged
        };
        this.openLoginDialog(data, 'adf-login-dialog', '630px');
        return logged;
    };
    /**
     * @param {?} data
     * @param {?} currentPanelClass
     * @param {?} chosenWidth
     * @return {?}
     */
    LoginDialogService.prototype.openLoginDialog = /**
     * @param {?} data
     * @param {?} currentPanelClass
     * @param {?} chosenWidth
     * @return {?}
     */
    function (data, currentPanelClass, chosenWidth) {
        this.dialog.open(LoginDialogComponent, { data: data, panelClass: currentPanelClass, width: chosenWidth });
    };
    /** Closes the currently open dialog. */
    /**
     * Closes the currently open dialog.
     * @return {?}
     */
    LoginDialogService.prototype.close = /**
     * Closes the currently open dialog.
     * @return {?}
     */
    function () {
        this.dialog.closeAll();
    };
    LoginDialogService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    LoginDialogService.ctorParameters = function () { return [
        { type: MatDialog }
    ]; };
    return LoginDialogService;
}());
export { LoginDialogService };
if (false) {
    /** @type {?} */
    LoginDialogService.prototype.dialog;
}
