// @alfresco/adf-core/esm5/services/page-title.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { AppConfigService } from '../app-config/app-config.service';
import { TranslationService } from './translation.service';
var PageTitleService = /** @class */ (function () {
    function PageTitleService(titleService, appConfig, translationService) {
        var _this = this;
        this.titleService = titleService;
        this.appConfig = appConfig;
        this.translationService = translationService;
        this.originalTitle = '';
        this.translatedTitle = '';
        translationService.translate.onLangChange.subscribe(function () { return _this.onLanguageChanged(); });
        translationService.translate.onTranslationChange.subscribe(function () { return _this.onLanguageChanged(); });
    }
    /**
     * Sets the page title.
     * @param value The new title
     */
    /**
     * Sets the page title.
     * @param {?=} value The new title
     * @return {?}
     */
    PageTitleService.prototype.setTitle = /**
     * Sets the page title.
     * @param {?=} value The new title
     * @return {?}
     */
    function (value) {
        if (value === void 0) { value = ''; }
        this.originalTitle = value;
        this.translatedTitle = this.translationService.instant(value);
        this.updateTitle();
    };
    /**
     * @return {?}
     */
    PageTitleService.prototype.onLanguageChanged = /**
     * @return {?}
     */
    function () {
        this.translatedTitle = this.translationService.instant(this.originalTitle);
        this.updateTitle();
    };
    /**
     * @return {?}
     */
    PageTitleService.prototype.updateTitle = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var name = this.appConfig.get('application.name') || 'Alfresco ADF Application';
        /** @type {?} */
        var title = this.translatedTitle ? this.translatedTitle + " - " + name : "" + name;
        this.titleService.setTitle(title);
    };
    PageTitleService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PageTitleService.ctorParameters = function () { return [
        { type: Title },
        { type: AppConfigService },
        { type: TranslationService }
    ]; };
    return PageTitleService;
}());
export { PageTitleService };
if (false) {
    /** @type {?} */
    PageTitleService.prototype.originalTitle;
    /** @type {?} */
    PageTitleService.prototype.translatedTitle;
    /** @type {?} */
    PageTitleService.prototype.titleService;
    /** @type {?} */
    PageTitleService.prototype.appConfig;
    /** @type {?} */
    PageTitleService.prototype.translationService;
}
