// @alfresco/adf-core/esm2015/card-view/components/card-view-boolitem/card-view-boolitem.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, Input } from '@angular/core';
import { CardViewBoolItemModel } from '../../models/card-view-boolitem.model';
import { CardViewUpdateService } from '../../services/card-view-update.service';
export class CardViewBoolItemComponent {
    /**
     * @param {?} cardViewUpdateService
     */
    constructor(cardViewUpdateService) {
        this.cardViewUpdateService = cardViewUpdateService;
    }
    /**
     * @return {?}
     */
    isEditable() {
        return this.editable && this.property.editable;
    }
    /**
     * @param {?} change
     * @return {?}
     */
    changed(change) {
        this.cardViewUpdateService.update(this.property, change.checked);
        this.property.value = change.checked;
    }
}
CardViewBoolItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-card-view-boolitem',
                template: "<ng-container *ngIf=\"!property.isEmpty() || isEditable()\">\n    <div [attr.data-automation-id]=\"'card-boolean-label-' + property.key\" class=\"adf-property-label\">{{ property.label | translate }}</div>\n    <div class=\"adf-property-value\">\n        <mat-checkbox\n            [attr.data-automation-id]=\"'card-boolean-' + property.key\"\n            [attr.title]=\"'CORE.METADATA.ACTIONS.TOGGLE' | translate\"\n            [checked]=\"property.displayValue\"\n            [disabled]=\"!isEditable()\"\n            (change)=\"changed($event)\">\n        </mat-checkbox>\n    </div>\n</ng-container>\n",
                styles: [""]
            }] }
];
/** @nocollapse */
CardViewBoolItemComponent.ctorParameters = () => [
    { type: CardViewUpdateService }
];
CardViewBoolItemComponent.propDecorators = {
    property: [{ type: Input }],
    editable: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    CardViewBoolItemComponent.prototype.property;
    /** @type {?} */
    CardViewBoolItemComponent.prototype.editable;
    /** @type {?} */
    CardViewBoolItemComponent.prototype.cardViewUpdateService;
}
