// @alfresco/adf-core/esm5/sorting-picker/sorting-picker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation, Input, EventEmitter, Output } from '@angular/core';
var SortingPickerComponent = /** @class */ (function () {
    function SortingPickerComponent() {
        /**
         * Available sorting options
         */
        this.options = [];
        /**
         * Current sorting direction
         */
        this.ascending = true;
        /**
         * Raised each time sorting key or direction gets changed.
         */
        this.change = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    SortingPickerComponent.prototype.onChanged = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.selected = event.value;
        this.raiseChangedEvent();
    };
    /**
     * @return {?}
     */
    SortingPickerComponent.prototype.toggleSortDirection = /**
     * @return {?}
     */
    function () {
        this.ascending = !this.ascending;
        this.raiseChangedEvent();
    };
    /**
     * @return {?}
     */
    SortingPickerComponent.prototype.raiseChangedEvent = /**
     * @return {?}
     */
    function () {
        this.change.emit({
            key: this.selected,
            ascending: this.ascending
        });
    };
    SortingPickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-sorting-picker',
                    template: "<mat-form-field>\n    <mat-select [(value)]=\"selected\" (selectionChange)=\"onChanged($event)\">\n        <mat-option *ngFor=\"let option of options\" [value]=\"option.key\">\n            {{ option.label | translate }}\n        </mat-option>\n    </mat-select>\n</mat-form-field>\n\n<button *ngIf=\"selected\" mat-icon-button (click)=\"toggleSortDirection()\">\n    <mat-icon *ngIf=\"ascending\">arrow_upward</mat-icon>\n    <mat-icon *ngIf=\"!ascending\">arrow_downward</mat-icon>\n</button>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: { class: 'adf-sorting-picker' }
                }] }
    ];
    SortingPickerComponent.propDecorators = {
        options: [{ type: Input }],
        selected: [{ type: Input }],
        ascending: [{ type: Input }],
        change: [{ type: Output }]
    };
    return SortingPickerComponent;
}());
export { SortingPickerComponent };
if (false) {
    /**
     * Available sorting options
     * @type {?}
     */
    SortingPickerComponent.prototype.options;
    /**
     * Currently selected option key
     * @type {?}
     */
    SortingPickerComponent.prototype.selected;
    /**
     * Current sorting direction
     * @type {?}
     */
    SortingPickerComponent.prototype.ascending;
    /**
     * Raised each time sorting key or direction gets changed.
     * @type {?}
     */
    SortingPickerComponent.prototype.change;
}
