// @alfresco/adf-content-services/esm2015/upload/components/upload-files.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export class UploadFilesEvent {
    /**
     * @param {?} files
     * @param {?} uploadService
     * @param {?} callback
     */
    constructor(files, uploadService, callback) {
        this.files = files;
        this.uploadService = uploadService;
        this.callback = callback;
        this.isDefaultPrevented = false;
    }
    /**
     * @return {?}
     */
    get defaultPrevented() {
        return this.isDefaultPrevented;
    }
    /**
     * @return {?}
     */
    preventDefault() {
        this.isDefaultPrevented = true;
    }
    /**
     * @return {?}
     */
    pauseUpload() {
        this.preventDefault();
    }
    /**
     * @return {?}
     */
    resumeUpload() {
        if (this.files && this.files.length > 0) {
            this.uploadService.addToQueue(...this.files);
            this.uploadService.uploadFilesInTheQueue(this.callback);
        }
    }
}
if (false) {
    /** @type {?} */
    UploadFilesEvent.prototype.isDefaultPrevented;
    /** @type {?} */
    UploadFilesEvent.prototype.files;
    /** @type {?} */
    UploadFilesEvent.prototype.uploadService;
    /** @type {?} */
    UploadFilesEvent.prototype.callback;
}
