// @alfresco/adf-core/esm5/language-menu/language-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component } from '@angular/core';
import { AppConfigService, AppConfigValues } from '../app-config/app-config.service';
import { UserPreferencesService } from '../services/user-preferences.service';
var LanguageMenuComponent = /** @class */ (function () {
    function LanguageMenuComponent(appConfig, userPreference) {
        this.appConfig = appConfig;
        this.userPreference = userPreference;
        this.languages = [
            { key: 'en', label: 'English' }
        ];
    }
    /**
     * @return {?}
     */
    LanguageMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var languagesCongifApp = this.appConfig.get(AppConfigValues.APP_CONFIG_LANGUAGES_KEY);
        if (languagesCongifApp) {
            this.languages = languagesCongifApp;
        }
    };
    /**
     * @param {?} lang
     * @return {?}
     */
    LanguageMenuComponent.prototype.changeLanguage = /**
     * @param {?} lang
     * @return {?}
     */
    function (lang) {
        this.userPreference.locale = lang;
    };
    LanguageMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-language-menu',
                    template: "<button mat-menu-item *ngFor=\"let language of languages\" (click)=\"changeLanguage(language.key)\">{{language.label}}\n</button>\n"
                }] }
    ];
    /** @nocollapse */
    LanguageMenuComponent.ctorParameters = function () { return [
        { type: AppConfigService },
        { type: UserPreferencesService }
    ]; };
    return LanguageMenuComponent;
}());
export { LanguageMenuComponent };
if (false) {
    /** @type {?} */
    LanguageMenuComponent.prototype.languages;
    /** @type {?} */
    LanguageMenuComponent.prototype.appConfig;
    /** @type {?} */
    LanguageMenuComponent.prototype.userPreference;
}
