// @alfresco/adf-core/esm2015/clipboard/clipboard.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Directive, Input, HostListener } from '@angular/core';
import { ClipboardService } from './clipboard.service';
export class ClipboardDirective {
    /**
     * @param {?} clipboardService
     */
    constructor(clipboardService) {
        this.clipboardService = clipboardService;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    handleClickEvent(event) {
        event.preventDefault();
        this.copyToClipboard();
    }
    /**
     * @return {?}
     */
    copyToClipboard() {
        /** @type {?} */
        const isValidTarget = this.clipboardService.isTargetValid(this.target);
        if (isValidTarget) {
            this.clipboardService.copyToClipboard(this.target, this.message);
        }
    }
}
ClipboardDirective.decorators = [
    { type: Directive, args: [{
                selector: '[adf-clipboard]',
                exportAs: 'adfClipboard'
            },] }
];
/** @nocollapse */
ClipboardDirective.ctorParameters = () => [
    { type: ClipboardService }
];
ClipboardDirective.propDecorators = {
    target: [{ type: Input, args: ['adf-clipboard',] }],
    message: [{ type: Input, args: ['clipboard-notification',] }],
    handleClickEvent: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    ClipboardDirective.prototype.target;
    /** @type {?} */
    ClipboardDirective.prototype.message;
    /** @type {?} */
    ClipboardDirective.prototype.clipboardService;
}
