// @alfresco/adf-core/esm2015/events/file.event.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { FileUploadStatus } from '../models/file.model';
export class FileUploadEvent {
    /**
     * @param {?} file
     * @param {?=} status
     * @param {?=} error
     */
    constructor(file, status = FileUploadStatus.Pending, error = null) {
        this.file = file;
        this.status = status;
        this.error = error;
    }
}
if (false) {
    /** @type {?} */
    FileUploadEvent.prototype.file;
    /** @type {?} */
    FileUploadEvent.prototype.status;
    /** @type {?} */
    FileUploadEvent.prototype.error;
}
export class FileUploadCompleteEvent extends FileUploadEvent {
    /**
     * @param {?} file
     * @param {?=} totalComplete
     * @param {?=} data
     * @param {?=} totalAborted
     */
    constructor(file, totalComplete = 0, data, totalAborted = 0) {
        super(file, FileUploadStatus.Complete);
        this.totalComplete = totalComplete;
        this.data = data;
        this.totalAborted = totalAborted;
    }
}
if (false) {
    /** @type {?} */
    FileUploadCompleteEvent.prototype.totalComplete;
    /** @type {?} */
    FileUploadCompleteEvent.prototype.data;
    /** @type {?} */
    FileUploadCompleteEvent.prototype.totalAborted;
}
export class FileUploadDeleteEvent extends FileUploadEvent {
    /**
     * @param {?} file
     * @param {?=} totalComplete
     */
    constructor(file, totalComplete = 0) {
        super(file, FileUploadStatus.Deleted);
        this.totalComplete = totalComplete;
    }
}
if (false) {
    /** @type {?} */
    FileUploadDeleteEvent.prototype.totalComplete;
}
export class FileUploadErrorEvent extends FileUploadEvent {
    /**
     * @param {?} file
     * @param {?} error
     * @param {?=} totalError
     */
    constructor(file, error, totalError = 0) {
        super(file, FileUploadStatus.Error);
        this.error = error;
        this.totalError = totalError;
    }
}
if (false) {
    /** @type {?} */
    FileUploadErrorEvent.prototype.error;
    /** @type {?} */
    FileUploadErrorEvent.prototype.totalError;
}
