// @alfresco/adf-core/esm5/toolbar/toolbar-divider.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation } from '@angular/core';
var ToolbarDividerComponent = /** @class */ (function () {
    function ToolbarDividerComponent() {
    }
    ToolbarDividerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-toolbar-divider',
                    template: '<div></div>',
                    host: { 'class': 'adf-toolbar-divider' },
                    encapsulation: ViewEncapsulation.None,
                    styles: ["\n        .adf-toolbar-divider > div {\n            height: 24px;\n            width: 1px;\n            background: rgba(0, 0, 0, 0.26);\n            margin-left: 5px;\n            margin-right: 5px;\n        }\n    "]
                }] }
    ];
    return ToolbarDividerComponent;
}());
export { ToolbarDividerComponent };
