// @alfresco/adf-content-services/esm5/permission-manager/services/node-permission-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { MatDialog } from '@angular/material';
import { Injectable } from '@angular/core';
import { Subject, throwError } from 'rxjs';
import { AddPermissionDialogComponent } from '../components/add-permission/add-permission-dialog.component';
import { NodePermissionService } from './node-permission.service';
import { ContentService, PermissionsEnum } from '@alfresco/adf-core';
import { switchMap } from 'rxjs/operators';
var NodePermissionDialogService = /** @class */ (function () {
    function NodePermissionDialogService(dialog, nodePermissionService, contentService) {
        this.dialog = dialog;
        this.nodePermissionService = nodePermissionService;
        this.contentService = contentService;
    }
    /**
     * Opens a dialog to add permissions to a node.
     * @param node ID of the target node
     * @param title Dialog title
     * @returns Node with updated permissions
     */
    /**
     * Opens a dialog to add permissions to a node.
     * @param {?} node ID of the target node
     * @param {?=} title Dialog title
     * @return {?} Node with updated permissions
     */
    NodePermissionDialogService.prototype.openAddPermissionDialog = /**
     * Opens a dialog to add permissions to a node.
     * @param {?} node ID of the target node
     * @param {?=} title Dialog title
     * @return {?} Node with updated permissions
     */
    function (node, title) {
        if (this.contentService.hasPermission(node, PermissionsEnum.UPDATEPERMISSIONS)) {
            /** @type {?} */
            var confirm_1 = new Subject();
            confirm_1.subscribe({
                complete: this.close.bind(this)
            });
            /** @type {?} */
            var data = {
                nodeId: node.id,
                title: title,
                confirm: confirm_1
            };
            this.openDialog(data, 'adf-add-permission-dialog', '630px');
            return confirm_1;
        }
        else {
            /** @type {?} */
            var errors = new Error(JSON.stringify({ error: { statusCode: 403 } }));
            errors.message = 'PERMISSION_MANAGER.ERROR.NOT-ALLOWED';
            return throwError(errors);
        }
    };
    /**
     * @param {?} data
     * @param {?} currentPanelClass
     * @param {?} chosenWidth
     * @return {?}
     */
    NodePermissionDialogService.prototype.openDialog = /**
     * @param {?} data
     * @param {?} currentPanelClass
     * @param {?} chosenWidth
     * @return {?}
     */
    function (data, currentPanelClass, chosenWidth) {
        this.dialog.open(AddPermissionDialogComponent, { data: data, panelClass: currentPanelClass, width: chosenWidth });
    };
    /**
     * Closes the currently-open dialog.
     */
    /**
     * Closes the currently-open dialog.
     * @return {?}
     */
    NodePermissionDialogService.prototype.close = /**
     * Closes the currently-open dialog.
     * @return {?}
     */
    function () {
        this.dialog.closeAll();
    };
    /**
     * Opens a dialog to update permissions for a node.
     * @param nodeId ID of the target node
     * @param title Dialog title
     * @returns Node with updated permissions
     */
    /**
     * Opens a dialog to update permissions for a node.
     * @param {?=} nodeId ID of the target node
     * @param {?=} title Dialog title
     * @return {?} Node with updated permissions
     */
    NodePermissionDialogService.prototype.updateNodePermissionByDialog = /**
     * Opens a dialog to update permissions for a node.
     * @param {?=} nodeId ID of the target node
     * @param {?=} title Dialog title
     * @return {?} Node with updated permissions
     */
    function (nodeId, title) {
        var _this = this;
        return this.contentService.getNode(nodeId, { include: ['allowableOperations'] })
            .pipe(switchMap(function (node) {
            return _this.openAddPermissionDialog(node.entry, title)
                .pipe(switchMap(function (selection) {
                return _this.nodePermissionService.updateNodePermissions(nodeId, selection);
            }));
        }));
    };
    NodePermissionDialogService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    NodePermissionDialogService.ctorParameters = function () { return [
        { type: MatDialog },
        { type: NodePermissionService },
        { type: ContentService }
    ]; };
    return NodePermissionDialogService;
}());
export { NodePermissionDialogService };
if (false) {
    /** @type {?} */
    NodePermissionDialogService.prototype.dialog;
    /** @type {?} */
    NodePermissionDialogService.prototype.nodePermissionService;
    /** @type {?} */
    NodePermissionDialogService.prototype.contentService;
}
