// @alfresco/adf-core/esm5/models/file.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @record
 */
export function FileUploadProgress() { }
/** @type {?} */
FileUploadProgress.prototype.loaded;
/** @type {?} */
FileUploadProgress.prototype.total;
/** @type {?} */
FileUploadProgress.prototype.percent;
/**
 * @record
 */
export function FileUploadOptions() { }
/** @type {?|undefined} */
FileUploadOptions.prototype.comment;
/** @type {?|undefined} */
FileUploadOptions.prototype.newVersion;
/** @type {?|undefined} */
FileUploadOptions.prototype.majorVersion;
/** @type {?|undefined} */
FileUploadOptions.prototype.parentId;
/** @type {?|undefined} */
FileUploadOptions.prototype.path;
/** @type {?|undefined} */
FileUploadOptions.prototype.nodeType;
/** @type {?|undefined} */
FileUploadOptions.prototype.properties;
/** @type {?|undefined} */
FileUploadOptions.prototype.association;
/** @type {?|undefined} */
FileUploadOptions.prototype.secondaryChildren;
/** @type {?|undefined} */
FileUploadOptions.prototype.targets;
/** @enum {number} */
var FileUploadStatus = {
    Pending: 0,
    Complete: 1,
    Starting: 2,
    Progress: 3,
    Cancelled: 4,
    Aborted: 5,
    Error: 6,
    Deleted: 7,
};
export { FileUploadStatus };
FileUploadStatus[FileUploadStatus.Pending] = 'Pending';
FileUploadStatus[FileUploadStatus.Complete] = 'Complete';
FileUploadStatus[FileUploadStatus.Starting] = 'Starting';
FileUploadStatus[FileUploadStatus.Progress] = 'Progress';
FileUploadStatus[FileUploadStatus.Cancelled] = 'Cancelled';
FileUploadStatus[FileUploadStatus.Aborted] = 'Aborted';
FileUploadStatus[FileUploadStatus.Error] = 'Error';
FileUploadStatus[FileUploadStatus.Deleted] = 'Deleted';
var FileModel = /** @class */ (function () {
    function FileModel(file, options, id) {
        this.status = FileUploadStatus.Pending;
        this.file = file;
        this.id = id;
        this.name = file.name;
        this.size = file.size;
        this.data = null;
        this.progress = {
            loaded: 0,
            total: 0,
            percent: 0
        };
        this.options = Object.assign({}, {
            newVersion: false
        }, options);
    }
    Object.defineProperty(FileModel.prototype, "extension", {
        get: /**
         * @return {?}
         */
        function () {
            return this.name.slice((Math.max(0, this.name.lastIndexOf('.')) || Infinity) + 1);
        },
        enumerable: true,
        configurable: true
    });
    return FileModel;
}());
export { FileModel };
if (false) {
    /** @type {?} */
    FileModel.prototype.name;
    /** @type {?} */
    FileModel.prototype.size;
    /** @type {?} */
    FileModel.prototype.file;
    /** @type {?} */
    FileModel.prototype.id;
    /** @type {?} */
    FileModel.prototype.status;
    /** @type {?} */
    FileModel.prototype.progress;
    /** @type {?} */
    FileModel.prototype.options;
    /** @type {?} */
    FileModel.prototype.data;
}
