// @alfresco/adf-core/esm5/models/pagination.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var PaginationModel = /** @class */ (function () {
    function PaginationModel(obj) {
        if (obj) {
            this.count = obj.count;
            this.hasMoreItems = obj.hasMoreItems ? obj.hasMoreItems : false;
            this.merge = obj.merge ? obj.merge : false;
            this.totalItems = obj.totalItems;
            this.skipCount = obj.skipCount;
            this.maxItems = obj.maxItems;
        }
    }
    return PaginationModel;
}());
export { PaginationModel };
if (false) {
    /** @type {?} */
    PaginationModel.prototype.count;
    /** @type {?} */
    PaginationModel.prototype.hasMoreItems;
    /** @type {?} */
    PaginationModel.prototype.merge;
    /** @type {?} */
    PaginationModel.prototype.totalItems;
    /** @type {?} */
    PaginationModel.prototype.skipCount;
    /** @type {?} */
    PaginationModel.prototype.maxItems;
}
