// @alfresco/adf-core/esm5/userinfo/services/ecm-user.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { from, throwError } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { ContentService } from '../../services/content.service';
import { AlfrescoApiService } from '../../services/alfresco-api.service';
import { LogService } from '../../services/log.service';
var EcmUserService = /** @class */ (function () {
    function EcmUserService(apiService, contentService, logService) {
        this.apiService = apiService;
        this.contentService = contentService;
        this.logService = logService;
    }
    /**
     * Gets information about a user identified by their username.
     * @param userName Target username
     * @returns User information
     */
    /**
     * Gets information about a user identified by their username.
     * @param {?} userName Target username
     * @return {?} User information
     */
    EcmUserService.prototype.getUserInfo = /**
     * Gets information about a user identified by their username.
     * @param {?} userName Target username
     * @return {?} User information
     */
    function (userName) {
        var _this = this;
        return from(this.apiService.getInstance().core.peopleApi.getPerson(userName))
            .pipe(map(function (data) { return (data['entry']); }), catchError(function (err) { return _this.handleError(err); }));
    };
    /**
     * Gets information about the user who is currently logged-in.
     * @returns User information as for getUserInfo
     */
    /**
     * Gets information about the user who is currently logged-in.
     * @return {?} User information as for getUserInfo
     */
    EcmUserService.prototype.getCurrentUserInfo = /**
     * Gets information about the user who is currently logged-in.
     * @return {?} User information as for getUserInfo
     */
    function () {
        return this.getUserInfo('-me-');
    };
    /**
     * Returns a profile image as a URL.
     * @param avatarId Target avatar
     * @returns Image URL
     */
    /**
     * Returns a profile image as a URL.
     * @param {?} avatarId Target avatar
     * @return {?} Image URL
     */
    EcmUserService.prototype.getUserProfileImage = /**
     * Returns a profile image as a URL.
     * @param {?} avatarId Target avatar
     * @return {?} Image URL
     */
    function (avatarId) {
        if (avatarId) {
            /** @type {?} */
            var nodeObj = { entry: { id: avatarId } };
            return this.contentService.getContentUrl(nodeObj);
        }
    };
    /**
     * Throw the error
     * @param {?} error
     * @return {?}
     */
    EcmUserService.prototype.handleError = /**
     * Throw the error
     * @param {?} error
     * @return {?}
     */
    function (error) {
        this.logService.error(error);
        return throwError(error || 'Server error');
    };
    EcmUserService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EcmUserService.ctorParameters = function () { return [
        { type: AlfrescoApiService },
        { type: ContentService },
        { type: LogService }
    ]; };
    return EcmUserService;
}());
export { EcmUserService };
if (false) {
    /** @type {?} */
    EcmUserService.prototype.apiService;
    /** @type {?} */
    EcmUserService.prototype.contentService;
    /** @type {?} */
    EcmUserService.prototype.logService;
}
