// @alfresco/adf-content-services/esm5/folder-directive/folder-create.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Directive, HostListener, Input, Output, EventEmitter } from '@angular/core';
import { MatDialog } from '@angular/material';
import { FolderDialogComponent } from '../dialogs/folder.dialog';
import { ContentService } from '@alfresco/adf-core';
/** @type {?} */
var DEFAULT_FOLDER_PARENT_ID = '-my-';
var FolderCreateDirective = /** @class */ (function () {
    function FolderCreateDirective(dialogRef, content) {
        this.dialogRef = dialogRef;
        this.content = content;
        /**
         * Parent folder where the new folder will be located after creation.
         */
        this.parentNodeId = DEFAULT_FOLDER_PARENT_ID;
        /**
         * Title of folder creation dialog.
         */
        this.title = null;
        /**
         * Type of node to create.
         */
        this.nodeType = 'cm:folder';
        /**
         * Emitted when an error occurs (eg, a folder with same name already exists).
         */
        this.error = new EventEmitter();
        /**
         * Emitted when the folder is created successfully.
         */
        this.success = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    FolderCreateDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        this.openDialog();
    };
    Object.defineProperty(FolderCreateDirective.prototype, "dialogConfig", {
        get: /**
         * @return {?}
         */
        function () {
            var width = FolderCreateDirective.DIALOG_WIDTH;
            var _a = this, parentNodeId = _a.parentNodeId, createTitle = _a.title, nodeType = _a.nodeType;
            return {
                data: { parentNodeId: parentNodeId, createTitle: createTitle, nodeType: nodeType },
                width: width + "px"
            };
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FolderCreateDirective.prototype.openDialog = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var _a = this, dialogRef = _a.dialogRef, dialogConfig = _a.dialogConfig, content = _a.content;
        /** @type {?} */
        var dialogInstance = dialogRef.open(FolderDialogComponent, dialogConfig);
        dialogInstance.componentInstance.error.subscribe(function (error) {
            _this.error.emit(error);
        });
        dialogInstance.componentInstance.success.subscribe(function (node) {
            _this.success.emit(node);
        });
        dialogInstance.afterClosed().subscribe(function (node) {
            if (node) {
                content.folderCreate.next(node);
            }
        });
    };
    FolderCreateDirective.DIALOG_WIDTH = 400;
    FolderCreateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[adf-create-folder]'
                },] }
    ];
    /** @nocollapse */
    FolderCreateDirective.ctorParameters = function () { return [
        { type: MatDialog },
        { type: ContentService }
    ]; };
    FolderCreateDirective.propDecorators = {
        parentNodeId: [{ type: Input, args: ['adf-create-folder',] }],
        title: [{ type: Input }],
        nodeType: [{ type: Input }],
        error: [{ type: Output }],
        success: [{ type: Output }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return FolderCreateDirective;
}());
export { FolderCreateDirective };
if (false) {
    /** @type {?} */
    FolderCreateDirective.DIALOG_WIDTH;
    /**
     * Parent folder where the new folder will be located after creation.
     * @type {?}
     */
    FolderCreateDirective.prototype.parentNodeId;
    /**
     * Title of folder creation dialog.
     * @type {?}
     */
    FolderCreateDirective.prototype.title;
    /**
     * Type of node to create.
     * @type {?}
     */
    FolderCreateDirective.prototype.nodeType;
    /**
     * Emitted when an error occurs (eg, a folder with same name already exists).
     * @type {?}
     */
    FolderCreateDirective.prototype.error;
    /**
     * Emitted when the folder is created successfully.
     * @type {?}
     */
    FolderCreateDirective.prototype.success;
    /** @type {?} */
    FolderCreateDirective.prototype.dialogRef;
    /** @type {?} */
    FolderCreateDirective.prototype.content;
}
