// @alfresco/adf-content-services/esm5/search/components/search-radio/search-radio.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation, Input } from '@angular/core';
import { SearchFilterList } from '../search-filter/models/search-filter-list.model';
/**
 * @record
 */
export function SearchRadioOption() { }
/** @type {?} */
SearchRadioOption.prototype.name;
/** @type {?} */
SearchRadioOption.prototype.value;
var SearchRadioComponent = /** @class */ (function () {
    function SearchRadioComponent() {
        this.pageSize = 5;
        this.options = new SearchFilterList();
    }
    /**
     * @return {?}
     */
    SearchRadioComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.settings) {
            this.pageSize = this.settings["pageSize"] || 5;
            if (this.settings["options"] && this.settings["options"].length > 0) {
                this.options = new SearchFilterList(this.settings["options"], this.pageSize);
            }
        }
        this.setValue(this.getSelectedValue());
    };
    /**
     * @return {?}
     */
    SearchRadioComponent.prototype.getSelectedValue = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var options = this.settings['options'] || [];
        if (options && options.length > 0) {
            /** @type {?} */
            var selected = options.find(function (opt) { return opt.default; });
            if (!selected) {
                selected = options[0];
            }
            return selected.value;
        }
        return null;
    };
    /**
     * @param {?} newValue
     * @return {?}
     */
    SearchRadioComponent.prototype.setValue = /**
     * @param {?} newValue
     * @return {?}
     */
    function (newValue) {
        this.value = newValue;
        this.context.queryFragments[this.id] = newValue;
        this.context.update();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    SearchRadioComponent.prototype.changeHandler = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.setValue(event.value);
    };
    SearchRadioComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-search-radio',
                    template: "<mat-radio-group\n    [(ngModel)]=\"value\"\n    (change)=\"changeHandler($event)\">\n    <mat-radio-button\n        *ngFor=\"let option of options\"\n        [value]=\"option.value\">\n        {{ option.name | translate }}\n    </mat-radio-button>\n</mat-radio-group>\n\n<div class=\"facet-buttons\" *ngIf=\"!options.fitsPage\">\n    <button mat-icon-button\n        *ngIf=\"options.canShowLessItems\"\n        title=\"{{ 'SEARCH.FILTER.ACTIONS.SHOW-LESS' | translate }}\"\n        (click)=\"options.showLessItems()\">\n        <mat-icon>keyboard_arrow_up</mat-icon>\n    </button>\n    <button mat-icon-button\n        *ngIf=\"options.canShowMoreItems\"\n        title=\"{{ 'SEARCH.FILTER.ACTIONS.SHOW-MORE' | translate }}\"\n        (click)=\"options.showMoreItems()\">\n        <mat-icon>keyboard_arrow_down</mat-icon>\n    </button>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: { class: 'adf-search-radio' },
                    styles: [".adf-search-radio .mat-radio-group{display:inline-flex;flex-direction:column}.adf-search-radio .mat-radio-button{margin:5px}"]
                }] }
    ];
    /** @nocollapse */
    SearchRadioComponent.ctorParameters = function () { return []; };
    SearchRadioComponent.propDecorators = {
        value: [{ type: Input }]
    };
    return SearchRadioComponent;
}());
export { SearchRadioComponent };
if (false) {
    /**
     * The value of the selected radio button.
     * @type {?}
     */
    SearchRadioComponent.prototype.value;
    /** @type {?} */
    SearchRadioComponent.prototype.id;
    /** @type {?} */
    SearchRadioComponent.prototype.settings;
    /** @type {?} */
    SearchRadioComponent.prototype.context;
    /** @type {?} */
    SearchRadioComponent.prototype.options;
    /** @type {?} */
    SearchRadioComponent.prototype.pageSize;
}
