// @alfresco/adf-core/esm2015/collapsable/accordion-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild, ViewEncapsulation } from '@angular/core';
import { MatExpansionPanel } from '@angular/material';
export class AccordionGroupComponent {
    constructor() {
        this._isOpen = false;
        this._isSelected = false;
        /**
         * Should the (expanded) accordion icon be shown?
         */
        this.hasAccordionIcon = true;
        /**
         * Emitted when the heading is clicked.
         */
        this.headingClick = new EventEmitter();
    }
    /**
     * Is this group currently open?
     * @param {?} value
     * @return {?}
     */
    set isOpen(value) {
        this._isOpen = value;
    }
    /**
     * @return {?}
     */
    get isOpen() {
        return this._isOpen;
    }
    /**
     * Is this group currently selected?
     * @param {?} value
     * @return {?}
     */
    set isSelected(value) {
        this._isSelected = value;
    }
    /**
     * @return {?}
     */
    get isSelected() {
        return this._isSelected;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.hasContent = this.contentWrapper.nativeElement && this.contentWrapper.nativeElement.children.length > 0;
    }
    /**
     * @return {?}
     */
    hasHeadingIcon() {
        return !!this.headingIcon;
    }
    /**
     * @return {?}
     */
    onHeaderClick() {
        this.headingClick.emit(this.heading);
    }
    /**
     * @return {?}
     */
    isExpandable() {
        if (!this.hasContent || !this.isOpen) {
            this.expandPanel();
        }
    }
    /**
     * @return {?}
     */
    expandPanel() {
        this.expansionPanel.expanded = !this.expansionPanel.expanded;
    }
    /**
     * @return {?}
     */
    toggleExpansion() {
        return this.isOpen && this.isSelected;
    }
}
AccordionGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-accordion-group',
                template: "<mat-accordion class=\"adf-panel\">\n    <mat-expansion-panel #expansionPanel\n        id=\"adf-expansion-panel-id\"\n        (click)=\"isExpandable()\"\n        [expanded]=\"toggleExpansion()\"\n        (opened)=\"onHeaderClick()\"\n        [hideToggle]=\"!hasAccordionIcon\">\n        <mat-expansion-panel-header>\n            <mat-panel-title>\n                <div class=\"adf-panel-heading\" data-automation-id=\"adf-panel-heading\" [ngClass]=\"{'adf-panel-heading-selected': isSelected}\">\n                    <div id=\"heading-icon\" *ngIf=\"hasHeadingIcon()\" class=\"adf-panel-heading-icon\">\n                        <mat-icon class=\"material-icons\"\n                            [matTooltip]=\"headingIconTooltip\"\n                            [matTooltipDisabled]=\"!headingIconTooltip\">\n                            {{headingIcon}}\n                        </mat-icon>\n                    </div>\n                    <div id=\"heading-text\" class=\"adf-panel-heading-text\">{{heading}}</div>\n                </div>\n            </mat-panel-title>\n        </mat-expansion-panel-header>\n        <div id=\"adf-expansion-panel-content-id\" #contentWrapper>\n            <ng-content></ng-content>\n        </div>\n    </mat-expansion-panel>\n</mat-accordion>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [""]
            }] }
];
/** @nocollapse */
AccordionGroupComponent.ctorParameters = () => [];
AccordionGroupComponent.propDecorators = {
    contentWrapper: [{ type: ViewChild, args: ['contentWrapper',] }],
    expansionPanel: [{ type: ViewChild, args: ['expansionPanel',] }],
    heading: [{ type: Input }],
    headingIcon: [{ type: Input }],
    headingIconTooltip: [{ type: Input }],
    hasAccordionIcon: [{ type: Input }],
    headingClick: [{ type: Output }],
    isOpen: [{ type: Input }],
    isSelected: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AccordionGroupComponent.prototype._isOpen;
    /** @type {?} */
    AccordionGroupComponent.prototype._isSelected;
    /** @type {?} */
    AccordionGroupComponent.prototype.contentWrapper;
    /** @type {?} */
    AccordionGroupComponent.prototype.expansionPanel;
    /**
     * Title heading for the group.
     * @type {?}
     */
    AccordionGroupComponent.prototype.heading;
    /**
     * The material design icon.
     * @type {?}
     */
    AccordionGroupComponent.prototype.headingIcon;
    /**
     * Tooltip message to be shown for headingIcon
     * @type {?}
     */
    AccordionGroupComponent.prototype.headingIconTooltip;
    /**
     * Should the (expanded) accordion icon be shown?
     * @type {?}
     */
    AccordionGroupComponent.prototype.hasAccordionIcon;
    /**
     * Emitted when the heading is clicked.
     * @type {?}
     */
    AccordionGroupComponent.prototype.headingClick;
    /** @type {?} */
    AccordionGroupComponent.prototype.hasContent;
}
