// @alfresco/adf-core/esm2015/form/components/widgets/container/container.widget.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ContainerModel } from './../core/container.model';
import { FormFieldTypes } from './../core/form-field-types';
export class ContainerWidgetComponentModel extends ContainerModel {
    /**
     * @param {?} field
     */
    constructor(field) {
        super(field);
        this.columns = [];
        this.isExpanded = true;
        this.rowspan = 1;
        this.colspan = 1;
        if (this.field) {
            this.columns = this.field.columns || [];
            this.isExpanded = !this.isCollapsedByDefault();
            this.colspan = field.colspan;
            this.rowspan = field.rowspan;
        }
    }
    /**
     * @return {?}
     */
    isGroup() {
        return this.type === FormFieldTypes.GROUP;
    }
    /**
     * @return {?}
     */
    isCollapsible() {
        /** @type {?} */
        let allowCollapse = false;
        if (this.isGroup() && this.field.params['allowCollapse']) {
            allowCollapse = /** @type {?} */ (this.field.params['allowCollapse']);
        }
        return allowCollapse;
    }
    /**
     * @return {?}
     */
    isCollapsedByDefault() {
        /** @type {?} */
        let collapseByDefault = false;
        if (this.isCollapsible() && this.field.params['collapseByDefault']) {
            collapseByDefault = /** @type {?} */ (this.field.params['collapseByDefault']);
        }
        return collapseByDefault;
    }
}
if (false) {
    /** @type {?} */
    ContainerWidgetComponentModel.prototype.columns;
    /** @type {?} */
    ContainerWidgetComponentModel.prototype.isExpanded;
    /** @type {?} */
    ContainerWidgetComponentModel.prototype.rowspan;
    /** @type {?} */
    ContainerWidgetComponentModel.prototype.colspan;
}
