// @alfresco/adf-core/esm2015/form/components/widgets/container/container.widget.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation } from '@angular/core';
import { FormService } from './../../../services/form.service';
import { baseHost, WidgetComponent } from './../widget.component';
import { ContainerWidgetComponentModel } from './container.widget.model';
export class ContainerWidgetComponent extends WidgetComponent {
    /**
     * @param {?} formService
     */
    constructor(formService) {
        super(formService);
        this.formService = formService;
    }
    /**
     * @return {?}
     */
    onExpanderClicked() {
        if (this.content && this.content.isCollapsible()) {
            this.content.isExpanded = !this.content.isExpanded;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.field) {
            this.content = new ContainerWidgetComponentModel(this.field);
        }
    }
    /**
     * Serializes column fields
     * @return {?}
     */
    get fields() {
        /** @type {?} */
        const fields = [];
        /** @type {?} */
        let rowContainsElement = true;
        /** @type {?} */
        let rowIndex = 0;
        while (rowContainsElement) {
            rowContainsElement = false;
            for (let i = 0; i < this.content.columns.length; i++) {
                /** @type {?} */
                let field = this.content.columns[i].fields[rowIndex];
                if (field) {
                    rowContainsElement = true;
                }
                fields.push(field);
            }
            rowIndex++;
        }
        return fields;
    }
    /**
     * Calculate the column width based on the numberOfColumns and current field's colspan property
     *
     * @param {?} field
     * @return {?}
     */
    getColumnWith(field) {
        /** @type {?} */
        const colspan = field ? field.colspan : 1;
        return (100 / this.content.json.numberOfColumns) * colspan + '%';
    }
}
ContainerWidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'container-widget',
                template: "<div [hidden]=\"!content?.isGroup()\" class=\"container-widget__header\">\n    <h4 class=\"container-widget__header-text\" id=\"container-header\"\n        [class.collapsible]=\"content?.isCollapsible()\">\n        <button *ngIf=\"content?.isCollapsible()\"\n                mat-icon-button\n                class=\"mdl-button--icon\"\n                (click)=\"onExpanderClicked()\">\n            <mat-icon>{{ content?.isExpanded ? 'expand_more' : 'expand_less' }}</mat-icon>\n        </button>\n        <span (click)=\"onExpanderClicked()\" id=\"container-header-label\">{{content.name}}</span>\n    </h4>\n</div>\n\n<section class=\"grid-list\" *ngIf=\"content?.isExpanded\">\n    <div class=\"grid-list-item\" *ngFor=\"let field of fields\" [style.width]=\"getColumnWith(field)\">\n        <adf-form-field *ngIf=\"field\" [field]=\"field\"></adf-form-field>\n    </div>\n</section>\n\n\n",
                host: baseHost,
                encapsulation: ViewEncapsulation.None,
                styles: [""]
            }] }
];
/** @nocollapse */
ContainerWidgetComponent.ctorParameters = () => [
    { type: FormService }
];
if (false) {
    /** @type {?} */
    ContainerWidgetComponent.prototype.content;
    /** @type {?} */
    ContainerWidgetComponent.prototype.formService;
}
