// @alfresco/adf-content-services/esm5/search/components/search-sorting-picker/search-sorting-picker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as tslib_1 from "tslib";
import { Component, ViewEncapsulation } from '@angular/core';
import { SearchQueryBuilderService } from '../../search-query-builder.service';
var SearchSortingPickerComponent = /** @class */ (function () {
    function SearchSortingPickerComponent(queryBuilder) {
        this.queryBuilder = queryBuilder;
        this.options = [];
    }
    /**
     * @return {?}
     */
    SearchSortingPickerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.options = this.queryBuilder.getSortingOptions();
        /** @type {?} */
        var primary = this.queryBuilder.getPrimarySorting();
        if (primary) {
            this.value = primary.key;
            this.ascending = primary.ascending;
        }
    };
    /**
     * @param {?} sorting
     * @return {?}
     */
    SearchSortingPickerComponent.prototype.onChanged = /**
     * @param {?} sorting
     * @return {?}
     */
    function (sorting) {
        this.value = sorting.key;
        this.ascending = sorting.ascending;
        this.applySorting();
    };
    /**
     * @param {?} key
     * @return {?}
     */
    SearchSortingPickerComponent.prototype.findOptionByKey = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (key) {
            return this.options.find(function (opt) { return opt.key === key; });
        }
        return null;
    };
    /**
     * @return {?}
     */
    SearchSortingPickerComponent.prototype.applySorting = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var option = this.findOptionByKey(this.value);
        if (option) {
            this.queryBuilder.sorting = [tslib_1.__assign({}, option, { ascending: this.ascending })];
            this.queryBuilder.update();
        }
    };
    SearchSortingPickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-search-sorting-picker',
                    template: "<adf-sorting-picker\n    [options]=\"options\"\n    [selected]=\"value\"\n    [ascending]=\"ascending\"\n    (change)=\"onChanged($event)\">\n</adf-sorting-picker>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: { class: 'adf-search-sorting-picker' }
                }] }
    ];
    /** @nocollapse */
    SearchSortingPickerComponent.ctorParameters = function () { return [
        { type: SearchQueryBuilderService }
    ]; };
    return SearchSortingPickerComponent;
}());
export { SearchSortingPickerComponent };
if (false) {
    /** @type {?} */
    SearchSortingPickerComponent.prototype.options;
    /** @type {?} */
    SearchSortingPickerComponent.prototype.value;
    /** @type {?} */
    SearchSortingPickerComponent.prototype.ascending;
    /** @type {?} */
    SearchSortingPickerComponent.prototype.queryBuilder;
}
