// @alfresco/adf-core/esm2015/datatable/components/datatable/date-cell.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation } from '@angular/core';
import { DataTableCellComponent } from './datatable-cell.component';
import { UserPreferencesService, UserPreferenceValues } from '../../../services/user-preferences.service';
export class DateCellComponent extends DataTableCellComponent {
    /**
     * @param {?} userPreferenceService
     */
    constructor(userPreferenceService) {
        super();
        if (userPreferenceService) {
            userPreferenceService
                .select(UserPreferenceValues.Locale)
                .subscribe(locale => {
                this.currentLocale = locale;
            });
        }
    }
    /**
     * @return {?}
     */
    get format() {
        if (this.column) {
            return this.column.format || 'medium';
        }
        return 'medium';
    }
}
DateCellComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-date-cell',
                template: `
        <ng-container>
            <span title="{{ tooltip | date:'medium' }}" *ngIf="format === 'timeAgo' else standard_date">
                {{ value | adfTimeAgo: currentLocale }}
            </span>
        </ng-container>
        <ng-template #standard_date>
            <span title="{{ tooltip | date:format }}">
                {{ value | date:format }}
            </span>
        </ng-template>
    `,
                encapsulation: ViewEncapsulation.None,
                host: { class: 'adf-date-cell' }
            }] }
];
/** @nocollapse */
DateCellComponent.ctorParameters = () => [
    { type: UserPreferencesService }
];
if (false) {
    /** @type {?} */
    DateCellComponent.prototype.currentLocale;
}
