// @alfresco/adf-core/esm5/layout/components/header/header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, Input, Output, EventEmitter, ViewEncapsulation } from '@angular/core';
var HeaderLayoutComponent = /** @class */ (function () {
    function HeaderLayoutComponent() {
        this.redirectUrl = '/';
        this.showSidenavToggle = true;
        this.clicked = new EventEmitter();
        /**
         * The side that the drawer is attached to 'start' | 'end' page
         */
        this.position = 'start';
    }
    /**
     * @return {?}
     */
    HeaderLayoutComponent.prototype.toggleMenu = /**
     * @return {?}
     */
    function () {
        this.clicked.emit(true);
    };
    /**
     * @return {?}
     */
    HeaderLayoutComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.logo) {
            this.logo = './assets/images/logo.png';
        }
    };
    HeaderLayoutComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-layout-header',
                    template: "<mat-toolbar color=\"{{color}}\" [style.background-color]=\"color\">\n    <button *ngIf=\"showSidenavToggle && position === 'start'\" id=\"adf-sidebar-toggle-start\"  data-automation-id=\"adf-menu-icon\"\n            class=\"mat-icon-button adf-menu-icon\" mat-icon-button (click)=\"toggleMenu()\">\n        <mat-icon class=\"mat-icon material-icon\" role=\"img\" aria-hidden=\"true\">menu</mat-icon>\n    </button>\n\n    <a [routerLink]=\"redirectUrl\" title=\"{{ tooltip }}\">\n        <img src=\"{{logo}}\" class=\"adf-app-logo\"/>\n    </a>\n\n    <span [routerLink]=\"redirectUrl\" fxFlex=\"1 1 auto\" fxShow fxHide.lt-sm=\"true\" class=\"adf-app-title\">{{title}}</span>\n    <ng-content></ng-content>\n\n    <button *ngIf=\"showSidenavToggle && position === 'end'\" id=\"adf-sidebar-toggle-end\" data-automation-id=\"adf-menu-icon\"\n            class=\"mat-icon-button adf-menu-icon\" mat-icon-button (click)=\"toggleMenu()\">\n        <mat-icon class=\"mat-icon material-icon\" role=\"img\" aria-hidden=\"true\">menu</mat-icon>\n    </button>\n\n</mat-toolbar>\n\n",
                    encapsulation: ViewEncapsulation.None,
                    host: { class: 'adf-layout-header' }
                }] }
    ];
    HeaderLayoutComponent.propDecorators = {
        title: [{ type: Input }],
        logo: [{ type: Input }],
        redirectUrl: [{ type: Input }],
        tooltip: [{ type: Input }],
        color: [{ type: Input }],
        showSidenavToggle: [{ type: Input }],
        clicked: [{ type: Output }],
        position: [{ type: Input }]
    };
    return HeaderLayoutComponent;
}());
export { HeaderLayoutComponent };
if (false) {
    /** @type {?} */
    HeaderLayoutComponent.prototype.title;
    /** @type {?} */
    HeaderLayoutComponent.prototype.logo;
    /** @type {?} */
    HeaderLayoutComponent.prototype.redirectUrl;
    /** @type {?} */
    HeaderLayoutComponent.prototype.tooltip;
    /** @type {?} */
    HeaderLayoutComponent.prototype.color;
    /** @type {?} */
    HeaderLayoutComponent.prototype.showSidenavToggle;
    /** @type {?} */
    HeaderLayoutComponent.prototype.clicked;
    /**
     * The side that the drawer is attached to 'start' | 'end' page
     * @type {?}
     */
    HeaderLayoutComponent.prototype.position;
}
