// @alfresco/adf-content-services/esm5/search/search-query-builder.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { AlfrescoApiService, AppConfigService } from '@alfresco/adf-core';
var SearchQueryBuilderService = /** @class */ (function () {
    function SearchQueryBuilderService(appConfig, alfrescoApiService) {
        this.appConfig = appConfig;
        this.alfrescoApiService = alfrescoApiService;
        this._userQuery = '';
        this.updated = new Subject();
        this.executed = new Subject();
        this.categories = [];
        this.queryFragments = {};
        this.filterQueries = [];
        this.paging = null;
        this.sorting = [];
        this.userFacetQueries = [];
        this.userFacetBuckets = {};
        // TODO: to be supported in future iterations
        this.ranges = {};
        this.resetToDefaults();
    }
    Object.defineProperty(SearchQueryBuilderService.prototype, "userQuery", {
        get: /**
         * @return {?}
         */
        function () {
            return this._userQuery;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            value = (value || '').trim();
            this._userQuery = value ? "(" + value + ")" : '';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SearchQueryBuilderService.prototype.resetToDefaults = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var template = this.appConfig.get('search');
        if (template) {
            this.config = JSON.parse(JSON.stringify(template));
            this.categories = (this.config.categories || []).filter(function (category) { return category.enabled; });
            this.filterQueries = this.config.filterQueries || [];
            this.userFacetBuckets = {};
            this.userFacetQueries = [];
            if (this.config.sorting) {
                this.sorting = this.config.sorting.defaults || [];
            }
        }
    };
    /**
     * @param {?} query
     * @return {?}
     */
    SearchQueryBuilderService.prototype.addUserFacetQuery = /**
     * @param {?} query
     * @return {?}
     */
    function (query) {
        if (query) {
            /** @type {?} */
            var existing = this.userFacetQueries.find(function (facetQuery) { return facetQuery.label === query.label; });
            if (existing) {
                existing.query = query.query;
            }
            else {
                this.userFacetQueries.push(tslib_1.__assign({}, query));
            }
        }
    };
    /**
     * @param {?} query
     * @return {?}
     */
    SearchQueryBuilderService.prototype.removeUserFacetQuery = /**
     * @param {?} query
     * @return {?}
     */
    function (query) {
        if (query) {
            this.userFacetQueries = this.userFacetQueries
                .filter(function (facetQuery) { return facetQuery.label !== query.label; });
        }
    };
    /**
     * @param {?} field
     * @param {?} bucket
     * @return {?}
     */
    SearchQueryBuilderService.prototype.addUserFacetBucket = /**
     * @param {?} field
     * @param {?} bucket
     * @return {?}
     */
    function (field, bucket) {
        if (field && field.field && bucket) {
            /** @type {?} */
            var buckets = this.userFacetBuckets[field.field] || [];
            /** @type {?} */
            var existing = buckets.find(function (facetBucket) { return facetBucket.label === bucket.label; });
            if (!existing) {
                buckets.push(bucket);
            }
            this.userFacetBuckets[field.field] = buckets;
        }
    };
    /**
     * @param {?} field
     * @return {?}
     */
    SearchQueryBuilderService.prototype.getUserFacetBuckets = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return this.userFacetBuckets[field] || [];
    };
    /**
     * @param {?} field
     * @param {?} bucket
     * @return {?}
     */
    SearchQueryBuilderService.prototype.removeUserFacetBucket = /**
     * @param {?} field
     * @param {?} bucket
     * @return {?}
     */
    function (field, bucket) {
        if (field && field.field && bucket) {
            /** @type {?} */
            var buckets = this.userFacetBuckets[field.field] || [];
            this.userFacetBuckets[field.field] = buckets
                .filter(function (facetBucket) { return facetBucket.label !== bucket.label; });
        }
    };
    /**
     * @param {?} query
     * @return {?}
     */
    SearchQueryBuilderService.prototype.addFilterQuery = /**
     * @param {?} query
     * @return {?}
     */
    function (query) {
        if (query) {
            /** @type {?} */
            var existing = this.filterQueries.find(function (filterQuery) { return filterQuery.query === query; });
            if (!existing) {
                this.filterQueries.push({ query: query });
            }
        }
    };
    /**
     * @param {?} query
     * @return {?}
     */
    SearchQueryBuilderService.prototype.removeFilterQuery = /**
     * @param {?} query
     * @return {?}
     */
    function (query) {
        if (query) {
            this.filterQueries = this.filterQueries
                .filter(function (filterQuery) { return filterQuery.query !== query; });
        }
    };
    /**
     * @param {?} label
     * @return {?}
     */
    SearchQueryBuilderService.prototype.getFacetQuery = /**
     * @param {?} label
     * @return {?}
     */
    function (label) {
        if (label && this.hasFacetQueries) {
            /** @type {?} */
            var result = this.config.facetQueries.queries.find(function (query) { return query.label === label; });
            if (result) {
                return tslib_1.__assign({}, result);
            }
        }
        return null;
    };
    /**
     * @param {?} label
     * @return {?}
     */
    SearchQueryBuilderService.prototype.getFacetField = /**
     * @param {?} label
     * @return {?}
     */
    function (label) {
        if (label) {
            /** @type {?} */
            var fields = this.config.facetFields.fields || [];
            /** @type {?} */
            var result = fields.find(function (field) { return field.label === label; });
            if (result) {
                return tslib_1.__assign({}, result);
            }
        }
        return null;
    };
    /**
     * @return {?}
     */
    SearchQueryBuilderService.prototype.update = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var query = this.buildQuery();
        this.updated.next(query);
    };
    /**
     * @return {?}
     */
    SearchQueryBuilderService.prototype.execute = /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var query, data;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        query = this.buildQuery();
                        if (!query) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.alfrescoApiService.searchApi.search(query)];
                    case 1:
                        data = _a.sent();
                        this.executed.next(data);
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @return {?}
     */
    SearchQueryBuilderService.prototype.buildQuery = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var query = this.getFinalQuery();
        /** @type {?} */
        var include = this.config.include || [];
        if (include.length === 0) {
            include.push('path', 'allowableOperations');
        }
        if (query) {
            /** @type {?} */
            var result = {
                query: {
                    query: query,
                    language: 'afts'
                },
                include: include,
                paging: this.paging,
                fields: this.config.fields,
                filterQueries: this.filterQueries,
                facetQueries: this.facetQueries,
                facetFields: this.facetFields,
                sort: this.sort
            };
            return result;
        }
        return null;
    };
    /**
     * Returns primary sorting definition.
     */
    /**
     * Returns primary sorting definition.
     * @return {?}
     */
    SearchQueryBuilderService.prototype.getPrimarySorting = /**
     * Returns primary sorting definition.
     * @return {?}
     */
    function () {
        if (this.sorting && this.sorting.length > 0) {
            return this.sorting[0];
        }
        return null;
    };
    /**
     * Returns all pre-configured sorting options that users can choose from.
     */
    /**
     * Returns all pre-configured sorting options that users can choose from.
     * @return {?}
     */
    SearchQueryBuilderService.prototype.getSortingOptions = /**
     * Returns all pre-configured sorting options that users can choose from.
     * @return {?}
     */
    function () {
        if (this.config && this.config.sorting) {
            return this.config.sorting.options || [];
        }
        return [];
    };
    Object.defineProperty(SearchQueryBuilderService.prototype, "hasFacetQueries", {
        /**
         * Check if FacetQueries has been defined
         */
        get: /**
         * Check if FacetQueries has been defined
         * @return {?}
         */
        function () {
            if (this.config
                && this.config.facetQueries
                && this.config.facetQueries.queries
                && this.config.facetQueries.queries.length > 0) {
                return true;
            }
            return false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SearchQueryBuilderService.prototype, "sort", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sorting.map(function (def) {
                return {
                    type: def.type,
                    field: def.field,
                    ascending: def.ascending
                };
            });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SearchQueryBuilderService.prototype, "facetQueries", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.hasFacetQueries) {
                return this.config.facetQueries.queries.map(function (query) {
                    return /** @type {?} */ (tslib_1.__assign({}, query));
                });
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SearchQueryBuilderService.prototype.getFinalQuery = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var query = '';
        this.categories.forEach(function (facet) {
            /** @type {?} */
            var customQuery = _this.queryFragments[facet.id];
            if (customQuery) {
                if (query.length > 0) {
                    query += ' AND ';
                }
                query += "(" + customQuery + ")";
            }
        });
        /** @type {?} */
        var result = [this.userQuery, query]
            .filter(function (entry) { return entry; })
            .join(' AND ');
        if (this.userFacetQueries && this.userFacetQueries.length > 0) {
            /** @type {?} */
            var combined = this.userFacetQueries
                .map(function (userQuery) { return userQuery.query; })
                .join(' OR ');
            result += " AND (" + combined + ")";
        }
        if (this.userFacetBuckets) {
            Object.keys(this.userFacetBuckets).forEach(function (key) {
                /** @type {?} */
                var subQuery = (_this.userFacetBuckets[key] || [])
                    .map(function (bucket) { return bucket.filterQuery; })
                    .join(' OR ');
                if (subQuery) {
                    if (result.length > 0) {
                        result += ' AND ';
                    }
                    result += "(" + subQuery + ")";
                }
            });
        }
        return result;
    };
    Object.defineProperty(SearchQueryBuilderService.prototype, "facetFields", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var facetFields = this.config.facetFields && this.config.facetFields.fields;
            if (facetFields && facetFields.length > 0) {
                return {
                    facets: facetFields.map(function (facet) { return ({
                        field: facet.field,
                        mincount: facet.mincount,
                        label: facet.label,
                        limit: facet.limit,
                        offset: facet.offset,
                        prefix: facet.prefix
                    }); })
                };
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    SearchQueryBuilderService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SearchQueryBuilderService.ctorParameters = function () { return [
        { type: AppConfigService },
        { type: AlfrescoApiService }
    ]; };
    return SearchQueryBuilderService;
}());
export { SearchQueryBuilderService };
if (false) {
    /** @type {?} */
    SearchQueryBuilderService.prototype._userQuery;
    /** @type {?} */
    SearchQueryBuilderService.prototype.updated;
    /** @type {?} */
    SearchQueryBuilderService.prototype.executed;
    /** @type {?} */
    SearchQueryBuilderService.prototype.categories;
    /** @type {?} */
    SearchQueryBuilderService.prototype.queryFragments;
    /** @type {?} */
    SearchQueryBuilderService.prototype.filterQueries;
    /** @type {?} */
    SearchQueryBuilderService.prototype.paging;
    /** @type {?} */
    SearchQueryBuilderService.prototype.sorting;
    /** @type {?} */
    SearchQueryBuilderService.prototype.userFacetQueries;
    /** @type {?} */
    SearchQueryBuilderService.prototype.userFacetBuckets;
    /** @type {?} */
    SearchQueryBuilderService.prototype.config;
    /** @type {?} */
    SearchQueryBuilderService.prototype.ranges;
    /** @type {?} */
    SearchQueryBuilderService.prototype.appConfig;
    /** @type {?} */
    SearchQueryBuilderService.prototype.alfrescoApiService;
}
