// @alfresco/adf-core/esm2015/services/login-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { MatDialog } from '@angular/material';
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { LoginDialogComponent } from '../login/components/login-dialog.component';
export class LoginDialogService {
    /**
     * @param {?} dialog
     */
    constructor(dialog) {
        this.dialog = dialog;
    }
    /**
     * Opens a dialog to choose a file to upload.
     * @param {?} actionName
     * @param {?} title
     * @return {?} Information about the chosen file(s)
     */
    openLogin(actionName, title) {
        /** @type {?} */
        const logged = new Subject();
        logged.subscribe({
            complete: this.close.bind(this)
        });
        /** @type {?} */
        const data = {
            title,
            actionName,
            logged
        };
        this.openLoginDialog(data, 'adf-login-dialog', '630px');
        return logged;
    }
    /**
     * @param {?} data
     * @param {?} currentPanelClass
     * @param {?} chosenWidth
     * @return {?}
     */
    openLoginDialog(data, currentPanelClass, chosenWidth) {
        this.dialog.open(LoginDialogComponent, { data, panelClass: currentPanelClass, width: chosenWidth });
    }
    /**
     * Closes the currently open dialog.
     * @return {?}
     */
    close() {
        this.dialog.closeAll();
    }
}
LoginDialogService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
LoginDialogService.ctorParameters = () => [
    { type: MatDialog }
];
if (false) {
    /** @type {?} */
    LoginDialogService.prototype.dialog;
}
