// @alfresco/adf-core/esm2015/services/page-title.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { AppConfigService } from '../app-config/app-config.service';
import { TranslationService } from './translation.service';
export class PageTitleService {
    /**
     * @param {?} titleService
     * @param {?} appConfig
     * @param {?} translationService
     */
    constructor(titleService, appConfig, translationService) {
        this.titleService = titleService;
        this.appConfig = appConfig;
        this.translationService = translationService;
        this.originalTitle = '';
        this.translatedTitle = '';
        translationService.translate.onLangChange.subscribe(() => this.onLanguageChanged());
        translationService.translate.onTranslationChange.subscribe(() => this.onLanguageChanged());
    }
    /**
     * Sets the page title.
     * @param {?=} value The new title
     * @return {?}
     */
    setTitle(value = '') {
        this.originalTitle = value;
        this.translatedTitle = this.translationService.instant(value);
        this.updateTitle();
    }
    /**
     * @return {?}
     */
    onLanguageChanged() {
        this.translatedTitle = this.translationService.instant(this.originalTitle);
        this.updateTitle();
    }
    /**
     * @return {?}
     */
    updateTitle() {
        /** @type {?} */
        const name = this.appConfig.get('application.name') || 'Alfresco ADF Application';
        /** @type {?} */
        const title = this.translatedTitle ? `${this.translatedTitle} - ${name}` : `${name}`;
        this.titleService.setTitle(title);
    }
}
PageTitleService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PageTitleService.ctorParameters = () => [
    { type: Title },
    { type: AppConfigService },
    { type: TranslationService }
];
if (false) {
    /** @type {?} */
    PageTitleService.prototype.originalTitle;
    /** @type {?} */
    PageTitleService.prototype.translatedTitle;
    /** @type {?} */
    PageTitleService.prototype.titleService;
    /** @type {?} */
    PageTitleService.prototype.appConfig;
    /** @type {?} */
    PageTitleService.prototype.translationService;
}
