// @alfresco/adf-core/esm2015/services/search.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Subject, from, throwError } from 'rxjs';
import { AlfrescoApiService } from './alfresco-api.service';
import { SearchConfigurationService } from './search-configuration.service';
import { catchError } from 'rxjs/operators';
export class SearchService {
    /**
     * @param {?} apiService
     * @param {?} searchConfigurationService
     */
    constructor(apiService, searchConfigurationService) {
        this.apiService = apiService;
        this.searchConfigurationService = searchConfigurationService;
        this.dataLoaded = new Subject();
    }
    /**
     * @param {?} term
     * @param {?=} options
     * @return {?}
     */
    getNodeQueryResults(term, options) {
        /** @type {?} */
        const promise = this.apiService.getInstance().core.queriesApi.findNodes(term, options);
        promise.then((data) => {
            this.dataLoaded.next(data);
        });
        return from(promise).pipe(catchError(err => this.handleError(err)));
    }
    /**
     * @param {?} searchTerm
     * @param {?} maxResults
     * @param {?} skipCount
     * @return {?}
     */
    search(searchTerm, maxResults, skipCount) {
        /** @type {?} */
        const searchQuery = Object.assign(this.searchConfigurationService.generateQueryBody(searchTerm, maxResults, skipCount));
        /** @type {?} */
        const promise = this.apiService.getInstance().search.searchApi.search(searchQuery);
        promise.then((data) => {
            this.dataLoaded.next(data);
        });
        return from(promise).pipe(catchError(err => this.handleError(err)));
    }
    /**
     * @param {?} queryBody
     * @return {?}
     */
    searchByQueryBody(queryBody) {
        /** @type {?} */
        const promise = this.apiService.getInstance().search.searchApi.search(queryBody);
        promise.then((data) => {
            this.dataLoaded.next(data);
        });
        return from(promise).pipe(catchError(err => this.handleError(err)));
    }
    /**
     * @param {?} error
     * @return {?}
     */
    handleError(error) {
        return throwError(error || 'Server error');
    }
}
SearchService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SearchService.ctorParameters = () => [
    { type: AlfrescoApiService },
    { type: SearchConfigurationService }
];
if (false) {
    /** @type {?} */
    SearchService.prototype.dataLoaded;
    /** @type {?} */
    SearchService.prototype.apiService;
    /** @type {?} */
    SearchService.prototype.searchConfigurationService;
}
/**
 * @record
 */
export function SearchOptions() { }
/** @type {?|undefined} */
SearchOptions.prototype.skipCount;
/** @type {?|undefined} */
SearchOptions.prototype.maxItems;
/** @type {?|undefined} */
SearchOptions.prototype.rootNodeId;
/** @type {?|undefined} */
SearchOptions.prototype.nodeType;
/** @type {?|undefined} */
SearchOptions.prototype.include;
/** @type {?|undefined} */
SearchOptions.prototype.orderBy;
/** @type {?|undefined} */
SearchOptions.prototype.fields;
