// @alfresco/adf-core/esm2015/services/settings.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { AppConfigService, AppConfigValues } from '../app-config/app-config.service';
import { LogService } from './log.service';
import { StorageService } from './storage.service';
export class SettingsService {
    /**
     * @param {?} appConfig
     * @param {?} logService
     * @param {?} storage
     */
    constructor(appConfig, logService, storage) {
        this.appConfig = appConfig;
        this.logService = logService;
        this.storage = storage;
    }
    /**
     * @deprecated in 1.6.0
     * @return {?}
     */
    get ecmHost() {
        this.logService.log('SettingsService.ecmHost is deprecated. Use AppConfigService instead.');
        return this.appConfig.get(AppConfigValues.ECMHOST);
    }
    /**
     * @deprecated in 1.7.0
     * @param {?} csrfDisabled
     * @return {?}
     */
    set csrfDisabled(csrfDisabled) {
        this.logService.log(`SettingsService.csrfDisabled is deprecated. Use UserPreferencesService.disableCSRF instead.`);
        if (csrfDisabled !== null && csrfDisabled !== undefined) {
            this.storage.setItem(AppConfigValues.DISABLECSRF, csrfDisabled.toString());
        }
    }
    /**
     * @deprecated in 1.6.0
     * @param {?} ecmHostUrl
     * @return {?}
     */
    set ecmHost(ecmHostUrl) {
        this.logService.log('SettingsService.ecmHost is deprecated. Use AppConfigService instead.');
    }
    /**
     * @deprecated in 1.6.0
     * @return {?}
     */
    get bpmHost() {
        this.logService.log('SettingsService.bpmHost is deprecated. Use AppConfigService instead.');
        return this.appConfig.get(AppConfigValues.BPMHOST);
    }
    /**
     * @deprecated in 1.6.0
     * @param {?} bpmHostUrl
     * @return {?}
     */
    set bpmHost(bpmHostUrl) {
        this.logService.log('SettingsService.bpmHost is deprecated. Use AppConfigService instead.');
    }
    /**
     * @deprecated in 1.6.0
     * @return {?}
     */
    getBPMApiBaseUrl() {
        this.logService.log('SettingsService.getBPMApiBaseUrl is deprecated.');
        return this.bpmHost + '/activiti-app';
    }
    /**
     * @deprecated in 1.7.0
     * @return {?}
     */
    getProviders() {
        this.logService.log(`SettingsService.getProviders is deprecated. Use UserPreferencesService.authType instead.`);
        return this.storage.getItem(AppConfigValues.PROVIDERS) || this.appConfig.get(AppConfigValues.PROVIDERS);
    }
    /**
     * @deprecated in 1.7.0
     * @param {?} providers
     * @return {?}
     */
    setProviders(providers) {
        this.logService.log(`SettingsService.setProviders is deprecated. Use the app-config.json`);
        if (providers) {
            this.storage.setItem(AppConfigValues.PROVIDERS, providers);
        }
    }
}
SettingsService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
SettingsService.ctorParameters = () => [
    { type: AppConfigService },
    { type: LogService },
    { type: StorageService }
];
if (false) {
    /** @type {?} */
    SettingsService.prototype.appConfig;
    /** @type {?} */
    SettingsService.prototype.logService;
    /** @type {?} */
    SettingsService.prototype.storage;
}
