// @alfresco/adf-core/esm5/card-view/models/card-view-baseitem.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
CardViewBaseItemModel = /** @class */ (function () {
    function CardViewBaseItemModel(obj) {
        this.label = obj.label || '';
        this.value = obj.value;
        this.key = obj.key;
        this.default = obj.default;
        this.editable = !!obj.editable;
        this.clickable = !!obj.clickable;
        this.icon = obj.icon || '';
        this.validators = obj.validators || [];
        this.data = obj.data || null;
    }
    /**
     * @return {?}
     */
    CardViewBaseItemModel.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return this.value === undefined || this.value === null || this.value === '';
    };
    /**
     * @param {?} newValue
     * @return {?}
     */
    CardViewBaseItemModel.prototype.isValid = /**
     * @param {?} newValue
     * @return {?}
     */
    function (newValue) {
        if (!this.validators.length) {
            return true;
        }
        return this.validators
            .map(function (validator) { return validator.isValid(newValue); })
            .reduce(function (isValidUntilNow, isValid) { return isValidUntilNow && isValid; }, true);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    CardViewBaseItemModel.prototype.getValidationErrors = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (!this.validators.length) {
            return [];
        }
        return this.validators.filter(function (validator) { return !validator.isValid(value); }).map(function (validator) { return validator.message; });
    };
    return CardViewBaseItemModel;
}());
/**
 * @abstract
 */
export { CardViewBaseItemModel };
if (false) {
    /** @type {?} */
    CardViewBaseItemModel.prototype.label;
    /** @type {?} */
    CardViewBaseItemModel.prototype.value;
    /** @type {?} */
    CardViewBaseItemModel.prototype.key;
    /** @type {?} */
    CardViewBaseItemModel.prototype.default;
    /** @type {?} */
    CardViewBaseItemModel.prototype.editable;
    /** @type {?} */
    CardViewBaseItemModel.prototype.clickable;
    /** @type {?} */
    CardViewBaseItemModel.prototype.icon;
    /** @type {?} */
    CardViewBaseItemModel.prototype.validators;
    /** @type {?} */
    CardViewBaseItemModel.prototype.data;
}
