// @alfresco/adf-core/esm5/form/components/widgets/core/form-widget.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
FormWidgetModel = /** @class */ (function () {
    function FormWidgetModel(form, json) {
        this.form = form;
        this.json = json;
        if (json) {
            this.fieldType = json.fieldType;
            this.id = json.id;
            this.name = json.name;
            this.type = json.type;
            this.tab = json.tab;
        }
    }
    return FormWidgetModel;
}());
/**
 * @abstract
 */
export { FormWidgetModel };
if (false) {
    /** @type {?} */
    FormWidgetModel.prototype.fieldType;
    /** @type {?} */
    FormWidgetModel.prototype.id;
    /** @type {?} */
    FormWidgetModel.prototype.name;
    /** @type {?} */
    FormWidgetModel.prototype.type;
    /** @type {?} */
    FormWidgetModel.prototype.tab;
    /** @type {?} */
    FormWidgetModel.prototype.form;
    /** @type {?} */
    FormWidgetModel.prototype.json;
}
// unsupported: template constraints.
/**
 * @record
 * @template T
 */
export function FormWidgetModelCache() { }
