// @alfresco/adf-content-services/esm5/content-metadata/services/config/content-metadata-config.factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { AppConfigService, LogService } from '@alfresco/adf-core';
import { AspectOrientedConfigService } from './aspect-oriented-config.service';
import { IndifferentConfigService } from './indifferent-config.service';
import { LayoutOrientedConfigService } from './layout-oriented-config.service';
var ContentMetadataConfigFactory = /** @class */ (function () {
    function ContentMetadataConfigFactory(appConfigService, logService) {
        this.appConfigService = appConfigService;
        this.logService = logService;
    }
    /**
     * @param {?=} presetName
     * @return {?}
     */
    ContentMetadataConfigFactory.prototype.get = /**
     * @param {?=} presetName
     * @return {?}
     */
    function (presetName) {
        if (presetName === void 0) { presetName = 'default'; }
        /** @type {?} */
        var presetConfig;
        try {
            presetConfig = this.appConfigService.config['content-metadata'].presets[presetName];
        }
        catch (_a) {
            if (presetName !== ContentMetadataConfigFactory.DEFAULT_PRESET_NAME) {
                this.logService.error("No content-metadata preset for: " + presetName);
            }
            presetConfig = ContentMetadataConfigFactory.INDIFFERENT_PRESET;
        }
        return this.createConfig(presetConfig);
    };
    /**
     * @param {?} presetConfig
     * @return {?}
     */
    ContentMetadataConfigFactory.prototype.createConfig = /**
     * @param {?} presetConfig
     * @return {?}
     */
    function (presetConfig) {
        /** @type {?} */
        var config;
        if (this.isLayoutOrientedPreset(presetConfig)) {
            config = new LayoutOrientedConfigService(/** @type {?} */ (presetConfig));
        }
        else if (this.isAspectOrientedPreset(presetConfig)) {
            config = new AspectOrientedConfigService(/** @type {?} */ (presetConfig));
        }
        else {
            config = new IndifferentConfigService(/** @type {?} */ (presetConfig));
        }
        Object.freeze(config);
        return config;
    };
    /**
     * @param {?} presetConfig
     * @return {?}
     */
    ContentMetadataConfigFactory.prototype.isAspectOrientedPreset = /**
     * @param {?} presetConfig
     * @return {?}
     */
    function (presetConfig) {
        return this.isObject(presetConfig);
    };
    /**
     * @param {?} presetConfig
     * @return {?}
     */
    ContentMetadataConfigFactory.prototype.isLayoutOrientedPreset = /**
     * @param {?} presetConfig
     * @return {?}
     */
    function (presetConfig) {
        return Array.isArray(presetConfig);
    };
    /**
     * @param {?} x
     * @return {?}
     */
    ContentMetadataConfigFactory.prototype.isObject = /**
     * @param {?} x
     * @return {?}
     */
    function (x) {
        return x != null && typeof x === 'object';
    };
    ContentMetadataConfigFactory.INDIFFERENT_PRESET = '*';
    ContentMetadataConfigFactory.DEFAULT_PRESET_NAME = 'default';
    ContentMetadataConfigFactory.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ContentMetadataConfigFactory.ctorParameters = function () { return [
        { type: AppConfigService },
        { type: LogService }
    ]; };
    return ContentMetadataConfigFactory;
}());
export { ContentMetadataConfigFactory };
if (false) {
    /** @type {?} */
    ContentMetadataConfigFactory.INDIFFERENT_PRESET;
    /** @type {?} */
    ContentMetadataConfigFactory.DEFAULT_PRESET_NAME;
    /** @type {?} */
    ContentMetadataConfigFactory.prototype.appConfigService;
    /** @type {?} */
    ContentMetadataConfigFactory.prototype.logService;
}
