// @alfresco/adf-core/esm2015/info-drawer/info-drawer-layout.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, Directive, ViewEncapsulation } from '@angular/core';
export class InfoDrawerLayoutComponent {
}
InfoDrawerLayoutComponent.decorators = [
    { type: Component, args: [{
                selector: 'adf-info-drawer-layout',
                template: "<div class=\"adf-info-drawer-layout-header\">\n    <div class=\"adf-info-drawer-layout-header-title\">\n        <ng-content select=\"[info-drawer-title]\"></ng-content>\n    </div>\n    <div class=\"adf-info-drawer-layout-header-buttons\">\n        <ng-content select=\"[info-drawer-buttons]\"></ng-content>\n    </div>\n</div>\n<div class=\"adf-info-drawer-layout-content\">\n    <ng-content select=\"[info-drawer-content]\"></ng-content>\n</div>",
                encapsulation: ViewEncapsulation.None,
                host: { 'class': 'adf-info-drawer-layout' },
                styles: [""]
            }] }
];
export class InfoDrawerTitleDirective {
}
InfoDrawerTitleDirective.decorators = [
    { type: Directive, args: [{ selector: '[info-drawer-title]' },] }
];
export class InfoDrawerButtonsDirective {
}
InfoDrawerButtonsDirective.decorators = [
    { type: Directive, args: [{ selector: '[info-drawer-buttons]' },] }
];
export class InfoDrawerContentDirective {
}
InfoDrawerContentDirective.decorators = [
    { type: Directive, args: [{ selector: '[info-drawer-content]' },] }
];
