// @alfresco/adf-content-services/esm5/permission-manager/components/inherited-button.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { NodesApiService, ContentService, PermissionsEnum } from '@alfresco/adf-core';
var InheritPermissionDirective = /** @class */ (function () {
    function InheritPermissionDirective(nodeService, contentService) {
        this.nodeService = nodeService;
        this.contentService = contentService;
        /**
         * Emitted when the node is updated.
         */
        this.updated = new EventEmitter();
        /**
         * Emitted when an error occurs.
         */
        this.error = new EventEmitter();
    }
    /**
     * @return {?}
     */
    InheritPermissionDirective.prototype.onInheritPermissionClicked = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.nodeService.getNode(this.nodeId).subscribe(function (node) {
            if (_this.contentService.hasPermission(node, PermissionsEnum.UPDATEPERMISSIONS)) {
                /** @type {?} */
                var nodeBody = { permissions: { isInheritanceEnabled: !node.permissions.isInheritanceEnabled } };
                _this.nodeService.updateNode(_this.nodeId, nodeBody, { include: ['permissions'] }).subscribe(function (nodeUpdated) {
                    _this.updated.emit(nodeUpdated);
                }, function (error) { return _this.error.emit(error); });
            }
            else {
                _this.error.emit('PERMISSION_MANAGER.ERROR.NOT-ALLOWED');
            }
        });
    };
    InheritPermissionDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'button[adf-inherit-permission], mat-button-toggle[adf-inherit-permission]',
                    host: {
                        'role': 'button',
                        '(click)': 'onInheritPermissionClicked()'
                    }
                },] }
    ];
    /** @nocollapse */
    InheritPermissionDirective.ctorParameters = function () { return [
        { type: NodesApiService },
        { type: ContentService }
    ]; };
    InheritPermissionDirective.propDecorators = {
        nodeId: [{ type: Input }],
        updated: [{ type: Output }],
        error: [{ type: Output }]
    };
    return InheritPermissionDirective;
}());
export { InheritPermissionDirective };
if (false) {
    /**
     * ID of the node to add/remove inherited permissions.
     * @type {?}
     */
    InheritPermissionDirective.prototype.nodeId;
    /**
     * Emitted when the node is updated.
     * @type {?}
     */
    InheritPermissionDirective.prototype.updated;
    /**
     * Emitted when an error occurs.
     * @type {?}
     */
    InheritPermissionDirective.prototype.error;
    /** @type {?} */
    InheritPermissionDirective.prototype.nodeService;
    /** @type {?} */
    InheritPermissionDirective.prototype.contentService;
}
