// @alfresco/adf-core/esm2015/form/components/widgets/amount/amount.widget.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Component, ViewEncapsulation } from '@angular/core';
import { FormService } from './../../../services/form.service';
import { baseHost, WidgetComponent } from './../widget.component';
export class AmountWidgetComponent extends WidgetComponent {
    /**
     * @param {?} formService
     */
    constructor(formService) {
        super(formService);
        this.formService = formService;
        this.currency = AmountWidgetComponent.DEFAULT_CURRENCY;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.field && this.field.currency) {
            this.currency = this.field.currency;
        }
    }
}
AmountWidgetComponent.DEFAULT_CURRENCY = '$';
AmountWidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'amount-widget',
                template: "<div class=\"adf-amount-widget__container adf-amount-widget {{field.className}}\" [class.adf-invalid]=\"!field.isValid\" [class.adf-readonly]=\"field.readOnly\">\n    <mat-form-field class=\"adf-amount-widget__input\">\n        <label class=\"adf-label\" [attr.for]=\"field.id\">{{field.name}}<span *ngIf=\"isRequired()\">*</span></label>\n        <span matPrefix class=\"adf-amount-widget__prefix-spacing\"> {{currency }}</span>\n        <input matInput\n                class=\"adf-amount-widget\"\n                type=\"text\"\n                [id]=\"field.id\"\n                [required]=\"isRequired()\"\n                [placeholder]=\"field.placeholder\"\n                [value]=\"field.value\"\n                [(ngModel)]=\"field.value\"\n                (ngModelChange)=\"onFieldChanged(field)\"\n                [disabled]=\"field.readOnly\">\n    </mat-form-field>\n    <error-widget [error]=\"field.validationSummary\" ></error-widget>\n    <error-widget *ngIf=\"isInvalidFieldRequired()\" required=\"{{ 'FORM.FIELD.REQUIRED' | translate }}\"></error-widget>\n</div>\n",
                host: baseHost,
                encapsulation: ViewEncapsulation.None,
                styles: [".adf-amount-widget{width:100%;vertical-align:baseline!important}.adf-amount-widget .mat-input-placeholder{margin-top:5px;display:none}.adf-amount-widget .mat-form-field-flex{position:relative;padding-top:18.5px}.adf-amount-widget .mat-form-field-infix{position:static;padding-top:19px}.adf-amount-widget .adf-label{position:absolute;top:18.5px;left:0}.adf-amount-widget__container{max-width:100%}.adf-amount-widget__container .mat-form-field-label-wrapper{top:34px!important;left:13px}.adf-amount-widget__input .mat-focused{transition:none}.adf-amount-widget__prefix-spacing{padding-right:5px}"]
            }] }
];
/** @nocollapse */
AmountWidgetComponent.ctorParameters = () => [
    { type: FormService }
];
if (false) {
    /** @type {?} */
    AmountWidgetComponent.DEFAULT_CURRENCY;
    /** @type {?} */
    AmountWidgetComponent.prototype.currency;
    /** @type {?} */
    AmountWidgetComponent.prototype.formService;
}
