// @alfresco/adf-content-services/esm5/breadcrumb/dropdown-breadcrumb.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/*!
 * @license
 * Copyright 2016 Alfresco Software, Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { MatSelect } from '@angular/material';
import { BreadcrumbComponent } from './breadcrumb.component';
var DropdownBreadcrumbComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DropdownBreadcrumbComponent, _super);
    function DropdownBreadcrumbComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    DropdownBreadcrumbComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes["folderNode"]) {
            /** @type {?} */
            var node = null;
            node = this.transform ? this.transform(changes["folderNode"].currentValue) : changes["folderNode"].currentValue;
            this.route = this.parseRoute(node);
        }
        if (changes["transform"]) {
            /** @type {?} */
            var node = this.transform ? this.transform(this.folderNode) : this.folderNode;
            this.route = this.parseRoute(node);
        }
        this.recalculateNodes();
    };
    /**
     * Calculate the current and previous nodes from the route array
     */
    /**
     * Calculate the current and previous nodes from the route array
     * @return {?}
     */
    DropdownBreadcrumbComponent.prototype.recalculateNodes = /**
     * Calculate the current and previous nodes from the route array
     * @return {?}
     */
    function () {
        this.currentNode = this.route[this.route.length - 1];
        this.previousNodes = this.route.slice(0, this.route.length - 1).reverse();
    };
    /**
     * Opens the node picker menu
     */
    /**
     * Opens the node picker menu
     * @return {?}
     */
    DropdownBreadcrumbComponent.prototype.open = /**
     * Opens the node picker menu
     * @return {?}
     */
    function () {
        if (this.dropdown) {
            this.dropdown.open();
        }
    };
    /**
     * Return if route has more than one element (means: we are not in the root directory)
     */
    /**
     * Return if route has more than one element (means: we are not in the root directory)
     * @return {?}
     */
    DropdownBreadcrumbComponent.prototype.hasPreviousNodes = /**
     * Return if route has more than one element (means: we are not in the root directory)
     * @return {?}
     */
    function () {
        return this.previousNodes.length > 0;
    };
    DropdownBreadcrumbComponent.decorators = [
        { type: Component, args: [{
                    selector: 'adf-dropdown-breadcrumb',
                    template: "<ng-container *ngIf=\"route.length > 0\">\n\n    <button\n        tabindex=\"0\"\n        class=\"adf-dropdown-breadcrumb-trigger\"\n        (click)=\"open()\"\n        data-automation-id=\"dropdown-breadcrumb-trigger\">\n        <mat-icon [class.isRoot]=\"!hasPreviousNodes()\">folder</mat-icon>\n    </button>\n    <mat-icon class=\"adf-dropdown-breadcrumb-item-chevron\">chevron_right</mat-icon>\n\n    <mat-select\n        #dropdown\n        *ngIf=\"hasPreviousNodes()\"\n        class=\"adf-dropdown-breadcrumb-path\"\n        tabindex=\"0\"\n        data-automation-id=\"dropdown-breadcrumb-path\" >\n\n        <mat-option\n            *ngFor=\"let node of previousNodes;\"\n            (click)=\"onRoutePathClick(node, $event)\"\n            class=\"adf-dropdown-breadcrumb-path-option\"\n            tabindex=\"0\"\n            data-automation-class=\"dropdown-breadcrumb-path-option\">\n            {{ node.name | translate }}\n        </mat-option>\n    </mat-select>\n\n    <span\n        class=\"adf-current-folder\"\n        [class.isRoot]=\"!hasPreviousNodes()\"\n        data-automation-id=\"current-folder\">{{ currentNode.name }}</span>\n</ng-container>\n",
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        'class': 'adf-dropdown-breadcrumb'
                    },
                    styles: [""]
                }] }
    ];
    DropdownBreadcrumbComponent.propDecorators = {
        dropdown: [{ type: ViewChild, args: ['dropdown',] }]
    };
    return DropdownBreadcrumbComponent;
}(BreadcrumbComponent));
export { DropdownBreadcrumbComponent };
if (false) {
    /** @type {?} */
    DropdownBreadcrumbComponent.prototype.dropdown;
    /** @type {?} */
    DropdownBreadcrumbComponent.prototype.currentNode;
    /** @type {?} */
    DropdownBreadcrumbComponent.prototype.previousNodes;
}
